---
author: "Alex Linz"
date: "August 7, 2018"
---
<style>
body {
text-align: justify}
</style>


# Metatranscriptomics reveals interactions between phototrophs and heterotrophs in freshwater

```{r, Setup, eval = T, echo = F, message = F, warning = F}
# Set up environment
path <- "C:/Users/Goose and Gander/Documents/"
path2 <- "C:/Users/Goose and Gander/"

library(ggplot2)
library(cowplot)
library(reshape2)
library(ggrepel)
library(raster)
library(DESeq2)

zscore <- function(counts){
  z <- (counts - mean(counts)) / sd(counts)
  return(z)
}

# Metatranscriptome data
snorm <- read.csv(paste(path, "geodes_data_tables/Sparkling_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
tnorm <- read.csv(paste(path, "geodes_data_tables/Trout_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
mnorm <- read.csv(paste(path, "geodes_data_tables/Mendota_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)

snorm <- snorm[, which(colnames(snorm) != "GEODES014.nonrRNA" & colnames(snorm) != "GEODES033.nonrRNA")]
tnorm <- tnorm[, which(colnames(tnorm) != "GEODES065.nonrRNA")]
mnorm <- mnorm[, which(colnames(mnorm) != "GEODES158.nonrRNA")]

mendota_key <- read.csv(paste(path, "geodes_data_tables/Mendota_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)
spark_key <- read.csv(paste(path, "geodes_data_tables/Sparkling_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)
trout_key <- read.csv(paste(path, "geodes_data_tables/Trout_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)

# Sample data
metadata <- read.csv(file = paste(path2, "Desktop/geodes/bioinformatics_workflow/R_processing/sample_metadata.csv", sep = ""), header = T)
enviro_data <- read.csv(paste(path2, "Desktop/geodes/environmental_data/for_humans/compiled_field_data_for_R.csv", sep = ""), header = T)
enviro_data_no_sonde <- enviro_data[which(is.na(enviro_data$pH) == F), ]
chemdata <- read.csv(paste(path2, "Desktop/geodes/environmental_data/GEODES_TNTP.csv", sep = ""), header = T)
chemdata$Lake <- NA
chemdata$Lake[grep("ME", chemdata$Sample)] <- "Mendota"
chemdata$Lake[grep("SP", chemdata$Sample)] <- "Sparkling"
chemdata$Lake[grep("TB", chemdata$Sample)] <- "Trout"

# Metagenome data
# metaG_reads <- read.table(paste(path, "geodes_data_tables/GEODES_metaG_ID90_2018-03-10.readcounts.txt", sep = ""), row.names = 1, sep = "\t")
# colnames(metaG_reads) <- c("GEODES005", "GEODES006", "GEODES057", "GEODES058", "GEODES117", "GEODES118", "GEODES165", "GEODES166", "GEODES167", "GEODES168")
# metaG_key <- read.table(paste(path, "geodes_data_tables/GEODES_metaG_genekey_2018-03-12.txt", sep = ""), sep = "\t", quote = "")
# colnames(metaG_key) <- c("Gene", "Genome", "Taxonomy", "Product")
# lakekey <- c("Sparkling", "Sparkling", "Trout", "Trout", "Mendota", "Mendota", "Sparkling2009", "Sparkling2009", "Sparkling2009", "Sparkling2009")
# metaG_reads <- sweep(metaG_reads, 2, colSums(metaG_reads), "/")

```

### Main text tables and figures.

**Table 1. Comparison of Sparkling Lake, Lake Mendota, and Trout Bog.** These three lakes were chosen for comparative metatranscriptomics because of their varying trophic statuses, extensive historical data, and previous microbial sampling. Data on surface area, maximum depth, and development on shoreline courtesy of NTL-LTER <lter.limnology.wisc.edu>. Temperature, dissolved oxygen, pH, and conductivity were measured using a HydroLab DS5x Sonde and are averaged over all sampling depths and timepoints for each lake. Chlorophyll and phycocyanin concentrations were measured from the integrated epilimnion samples using a methanol extraction protocol and averaged over all timepoints. Secchi depth was measured at the first timepoint for each lake. Bacterial production was quantified via C14-leucine incorporation and averaged over all timepoints. Due to thunderstorms the night of July 8th, the final 1AM timepoint in Sparkling Lake was collected on July 9th instead. 

|                         | Lake Mendota      | Trout Bog                | Sparkling Lake           |
| ----------------------- |:-----------------:|:------------------------:|:------------------------:|
| Surface area (km^2)     | 39.6              | 0.001                    | 0.637                    |
| Maximum depth (m)       | 25.3              | 7.9                      | 20                       |
| Trophic status          | Eutrophic         | Humic                    | Oligotrophic             |
| Location                | Madison, WI USA   | Boulder Junction, WI USA | Boulder Junction, WI USA |
| GPS Coordinates         | 43.1113, -89.4255 | 46.0412, -89.6861        | 46.0091, -89.6695        |
| Shoreline development   | High              | Low                      | Moderate                 |
| Epilimnion sampling depth| 0-7m             | 0-1.5m                   | 0-4m                     |
| Temperature (C)         | `r round(mean(enviro_data$Temperature[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$Temperature[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$Temperature[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
|Dissolved oxygen (mg/L)  | `r round(mean(enviro_data$DO[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$DO[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$DO[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| pH                      | `r round(mean(enviro_data$pH[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$pH[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$pH[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| Conductivity (uS/cm)    | `r round(mean(enviro_data$Conductivity[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$Conductivity[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$Conductivity[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| Total phosphorus (ug/L) | `r round(mean(chemdata$TP[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TP[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TP[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total nitrogen (ug/L)   | `r round(mean(chemdata$TN[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TN[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TN[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total dissolved phosphorus (ug/L) | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total dissolved nitrogen (ug/L)   | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Sparkling")]), 2)` |
| Chlorophyll (ug/L)      | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7 & enviro_data$Timepoint != 40 & enviro_data$Timepoint != 44)]), 2)` | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5 & enviro_data$Timepoint != 20)]), 2)` | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Phycocyanin (ug/L)      | `r round(mean(enviro_data$Phycocyanin[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7 & enviro_data$Timepoint != 40 & enviro_data$Timepoint != 44)]), 2)` | `r round(mean(enviro_data$Phycocyanin[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5 & enviro_data$Timepoint != 20)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Bacterial production (cpm)    | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Secchi depth (m)        | 4.8               | 1.1                      | 6.2                      |
| Sampling dates (2016)   | July 14-16        | July 8-10                | July 6-9                 |
| Sunrise/sunset time     | 5:32/20:35        | 5:18/20:49               | 5:17/20:50               |

```{r, table2, eval = F, echo = F}
totals <- rowSums(snorm)
top10 <- rownames(snorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
spark_key[match(top10, spark_key$Gene),]

totals <- rowSums(mnorm)
top10 <- rownames(mnorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
mendota_key[match(top10, mendota_key$Gene),]

totals <- rowSums(tnorm)
top10 <- rownames(tnorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
trout_key[match(top10, trout_key$Gene),]

```

**Table 2. Top 10 most expressed genes in each study site.** Coding regions from reference genomes and metagenome assemblies were clustered at 97% sequence similarity, and the longest coding region was chosen as the representative sequence. Metatranscriptomic reads were mapped to these representative sequences. The top 10 most expressed sequences from each lake, without any filters, are presented here. Annotations and classifications are derived from the sequence to which each read mapped.


| Lake Mendota      | Trout Bog                | Sparkling Lake           |
|:-----------------|:------------------------|:------------------------|
| Photosystem II P680 reaction center D1 protein (Cyanobacteria) | Photosystem II P680 reaction center D1 protein (unclassified) | Hypothetical protein (unclassified) |
| Photosystem II P680 reaction center D1 protein (unclassified) | Ribulose-bisphosphate carboxylase large chain (unclassified) | Hypothetical protein (unclassified) |
| Photosystem II P680 reaction center D1 protein (Cyanobacteria) | Photosystem II CP43 chlorophyll apoprotein (unclassified) | Photosystem II P680 reaction center D1 protein (Cyanobacteria) |
| Hypothetical protein (unclassified) | Putative beta-barrel porin2, OmpL-like (unclassified) | Photosystem II P680 reaction center D1 protein (unclassified) |
| Hypothetical protein (unclassified) | Photosystem II P680 reaction center D1 protein (Cyanobacteria) | Photosystem II P680 reaction center D1 protein (Chitinophagaceae) |
| Ribulose-bisphosphate carboxylase large chain (Cyanobacteria) | Photosystem II P700 chlorophyll a apoprotein  A2 (unclassified) | Hypothetical protein (unclassified) |
| PQQ-dependent dehydrogenase (LD28) | Photosystem II CP47 chlorophyll apoprotein (unclassified) | Hypothetical protein (unclassified) |
| Hypothetical protein (unclassified) | Hypothetical protein (unclassified) | Photosystem II P680 reaction center D1 protein (unclassified) |
| Hypothetical protein (Bdellovibrio) | Photosystem II P700 chlorophyll a apoprotein  A2 (unclassified) | Hypothetical protein (unclassified) |
| Ribulose-bisphosphate carboxylase large chain (Cyanobacteria) | Photosystem II P680 reaction center D1 protein (unclassified) | Hypothetical protein (unclassified) |


```{r, table3, eval = F, echo = F}
totals <- rowSums(snorm)
top10 <- rownames(snorm)[order(totals, decreasing = T)]
print <- spark_key[match(top10, spark_key$Gene),]
print <- print[grep("Cyano", print$Taxonomy, invert = T), ]
print <- print[grep("photo|Photo", print$Product, invert = T), ]
print <- print[grep("chloro|Chloro|ribulose-bisphosphate carboxylase", print$Product, invert = T), ]
print <- print[grep("hypothetical|Hypothetical|internal standard", print$Product, invert = T), ]
print[1:10,3:4]

totals <- rowSums(mnorm)
top10 <- rownames(mnorm)[order(totals, decreasing = T)]
print <- mendota_key[match(top10, mendota_key$Gene),]
print <- print[grep("Cyano", print$Taxonomy, invert = T), ]
print <- print[grep("photo|Photo", print$Product, invert = T), ]
print <- print[grep("chloro|Chloro|ribulose-bisphosphate carboxylase", print$Product, invert = T), ]
print <- print[grep("hypothetical|Hypothetical|internal standard", print$Product, invert = T), ]
print[1:10,3:4]

totals <- rowSums(tnorm)
top10 <- rownames(tnorm)[order(totals, decreasing = T)]
print <- trout_key[match(top10, trout_key$Gene),]
print <- print[grep("Cyano", print$Taxonomy, invert = T), ]
print <- print[grep("photo|Photo", print$Product, invert = T), ]
print <- print[grep("chloro|Chloro|ribulose-bisphosphate carboxylase", print$Product, invert = T), ]
print <- print[grep("hypothetical|Hypothetical|internal standard|unknown function", print$Product, invert = T), ]
print[1:10,3:4]

```
**Table 3. Top 10 most expressed annotated heterotrophic genes in each study site.** Coding regions from reference genomes and metagenome assemblies were clustered at 97% sequence similarity, and the longest coding region was chosen as the representative sequence. Metatranscriptomic reads were mapped to these representative sequences. The top 10 most expressed genes from each lake, filtered to exclude photosynthetic genes, phototrophic organisms, hypothetical genes, and unclassified genes, are presented here. Annotations and classifications are derived from the sequence to which each read mapped.

| Lake Mendota      | Trout Bog                | Sparkling Lake           |
|:-----------------|:------------------------|:------------------------|
| PQQ-dependent dehydrogenase (LD28) | Putative beta-barrel porin-2, OmpL-like (unclassified) | Chaperonin GroEL (Deltaproteobacteria) |
| Translation elongation factor TU (Bacteroidetes) | Ig-like domain group 3 (unclassified) | Fibronectin type III domain (unclassified) |
| YTV protein (Actinobacteria) | Ig-like domain group 1 (unclassified) | S-layer homology domain (Fimbriimonas ginsengisoli) |
| Translation elongation factor EF-G (acI-A6) | PEP-CTERM motif (unclassified) | Cytochrome c oxidase subunit 1 (unclassified) |
| DNA-directed RNA polymerase (acI-A6) | Cytochrome c oxidase subunit 1 (unclassified) | Cytochrome c oxidase subunit 2 (unclassified) |
| Probable sodium:solute symporter (LD12) | Cytochrome b6 (unclassified) | HYR domain, putative metal-binding (unclassified) |
| DNA-directed RNA polymerase (acI-A6) | Ig-like group 1 (acI-B) | F-type H+-transporting ATPase subunit beta (Chitinophagia) |
| Phycoerythrin alpha chain (unclassified) | S-layer homology domain (Fimbriimonas) | molecular chaperone HtpG (unclassified) |
| ABC-type sugar transport system (acI-B1) | Ribonucleoside-diphosphate reductase alpha chain (Pedosphaera parvula) | Glucose dehydrogenase (unclassified) |
| Translation elongation factor EF-G (acI-A6) | DNA-directed RNA polymerase subunit B (unclassified) | Lactocepin (unclassified) |



```{r, fig1, eval = T, echo = F}
# Process metagenome gene key to include a phylum column. Fix any weird formats.
# 
# metaG_key$Taxonomy <- gsub("Bacteria;", "", metaG_key$Taxonomy)
# metaG_key$Taxonomy <- gsub("Eukaryota;", "", metaG_key$Taxonomy)
# metaG_key$Phylum <- sapply(strsplit(as.character(metaG_key$Taxonomy),";"), `[`, 1)
# 
# metaG_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("None", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified unclassified", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("unclassified", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Unclassified ", "Unclassified", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("TM7", "Saccharibacteria", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Ignavibacteriae", "Ignavibacteria", metaG_key$Phylum)
# metaG_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", metaG_key$Phylum)
# metaG_key$Phylum[which(is.na(metaG_key$Phylum) == T)] <- "Unclassified"
# metaG_key$Phylum[grep("Blank", metaG_key$Phylum)] <- "Unclassified"
# 
# # Remove unclassified genes to save on RAM
# metaG_key <- metaG_key[which(metaG_key$Phylum != "Unclassified"),]
# 
# # Split metagenome read count tables by lake and prep for melting
# metaG_reads$Genes <- rownames(metaG_reads)
# metaG_reads <- metaG_reads[match(metaG_reads$Genes, metaG_key$Gene),]
# spark_metaG <- metaG_reads[,c(1,2, 11)]
# trout_metaG <- metaG_reads[,c(3,4, 11)]
# mendota_metaG <- metaG_reads[,c(5,6, 11)]
# 
# # Use melt to switch from wide to long format
# spark_metaG <- melt(spark_metaG)
# trout_metaG <- melt(trout_metaG)
# mendota_metaG <- melt(mendota_metaG)
# 
# # Add phylum column to each melted lake table
# mendota_metaG$Phylum <- metaG_key$Phylum[match(mendota_metaG$Genes, metaG_key$Gene)]
# trout_metaG$Phylum <- metaG_key$Phylum[match(trout_metaG$Genes, metaG_key$Gene)]
# spark_metaG$Phylum <- metaG_key$Phylum[match(spark_metaG$Genes, metaG_key$Gene)]
# 
# # Aggregate counts by phylum
# mendota_metaG_phyla <- aggregate(value ~ Phylum, data = mendota_metaG, mean)
# spark_metaG_phyla <- aggregate(value ~ Phylum, data = spark_metaG, mean)
# trout_metaG_phyla <- aggregate(value ~ Phylum, data = trout_metaG, mean)
# 
# # # Remove metaG datasets not needed any more to save RAM
# # rm(metaG_reads)
# # rm(mendota_metaG)
# # rm(trout_metaG)
# # rm(spark_metaG)
# # rm(metaG_key)
# 
# # Include a key for colorcoding phyla by broader classification
# 
# type_key <- data.frame(phylum = c("Actinobacteria", "Bacteroidetes", "Chloroflexi", "Crenarchaeota", "Cryptophyta", "Cyanobacteria", "Gemmatimonadetes", "Heterokonta", "Ignavibacteria", "Planctomycetes", "Proteobacteria", "Verrucomicrobia", "Viruses"), type = c("Bacteria", "Bacteria", "Bacteria", "Archaea", "Algae", "Bacteria", "Bacteria", "Algae", "Bacteria", "Bacteria", "Bacteria", "Bacteria", "Viruses"))


#Process Mendota expression data

# Similar to above, add phylum column and change weird formats
mendota_key$Taxonomy <- gsub("Bacteria;", "", mendota_key$Taxonomy)
mendota_key$Taxonomy <- gsub("Proteobacteria;", "", mendota_key$Taxonomy)
mendota_key$Taxonomy <- gsub("Eukaryota;", "", mendota_key$Taxonomy)
mendota_key$Phylum <- sapply(strsplit(as.character(mendota_key$Taxonomy),";"), `[`, 1)

mendota_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("None", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Unclassified ", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", mendota_key$Phylum)
mendota_key$Phylum[grep("Blank", mendota_key$Phylum)] <- "Unclassified"

# # Remove unclassified genes and the internal standard
# mendota_key <- mendota_key[grep("Unclassified|internal standard", mendota_key$Phylum, invert = T), ]
# 
# # Prep read count table and switch to long format
# mnorm$Genes <- rownames(mnorm)
# mnorm <- mnorm[which(mnorm$Genes %in% mendota_key$Gene), ]
# colnames(mnorm) <- gsub(".nonrRNA", "", colnames(mnorm))
# mnorm <- melt(mnorm)
# # Add columns of timepoint and phylum
# mnorm$Timepoint <- metadata$Timepoint[match(mnorm$variable, metadata$Sample)]
# mnorm$Taxonomy <- mendota_key$Phylum[match(mnorm$Genes, mendota_key$Gene)]
# 
# # Aggregate by phylum
# mendota_phyla <- aggregate(value ~ Taxonomy, data = mnorm, mean)
# 
# # Keep the top half of both expressed and abundant phyla for plotting
# keep_phyla <- c(mendota_metaG_phyla$Phylum[which(mendota_metaG_phyla$value > quantile(mendota_metaG_phyla$value)[3])], mendota_phyla$Taxonomy[which(mendota_phyla$value > quantile(mendota_phyla$value)[3])])
# mendota_phyla <- mendota_phyla[which(mendota_phyla$Taxonomy %in% keep_phyla),]
# mendota_phyla$metaG <- mendota_metaG_phyla$value[match(mendota_phyla$Taxonomy, mendota_metaG_phyla$Phylum)]
# mendota_phyla$type <- type_key$type[match(mendota_phyla$Taxonomy, type_key$phylum)]
# mendota_phyla <- mendota_phyla[which(is.na(mendota_phyla$metaG) == F),]
# 
# # Color key: c("limegreen", "lavenderblush4", "royalblue", "goldenrod")) == algae, archaea, bacteria, viruses
# p1 <- ggplot(mendota_phyla, aes(x = metaG, y = value, color = type)) + geom_point(size = 2.5) + geom_label_repel(aes(label = Taxonomy, color = type), force = 5, size = 7.5) + scale_color_manual(values = c("royalblue", "goldenrod")) + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean Transcripts/L assigned", title = "Lake Mendota") + theme(legend.position = "none")
# 
# #Clear mendota data from workspace and start with sparkling
# 
# # rm(averaged_tax)
# # rm(mendota_intersect)
# # rm(wide_mnorm)
# # rm(mendota_metaG_phyla)
# # rm(mendota_phyla)
# 
# #Process Sparkling expression data
# 
# # Similar to above, add phylum column and change weird formats
spark_key$Taxonomy <- gsub("Bacteria;", "", spark_key$Taxonomy)
spark_key$Taxonomy <- gsub("Proteobacteria;", "", spark_key$Taxonomy)
spark_key$Taxonomy <- gsub("Eukaryota;", "", spark_key$Taxonomy)
spark_key$Phylum <- sapply(strsplit(as.character(spark_key$Taxonomy),";"), `[`, 1)

spark_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("None", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified unclassified", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("Unclassified ", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", spark_key$Phylum)
spark_key$Phylum[grep("Blank", spark_key$Phylum)] <- "Unclassified"

# # Remove unclassified genes and the internal standard
# spark_key <- spark_key[grep("Unclassified|internal standard", spark_key$Phylum, invert = T), ]
# 
# # Prep read count table and switch to long format
# snorm$Genes <- rownames(snorm)
# snorm <- snorm[which(snorm$Genes %in% spark_key$Gene), ]
# colnames(snorm) <- gsub(".nonrRNA", "", colnames(snorm))
# snorm <- melt(snorm)
# # Add columns of timepoint and phylum
# snorm$Timepoint <- metadata$Timepoint[match(snorm$variable, metadata$Sample)]
# snorm$Taxonomy <- spark_key$Phylum[match(snorm$Genes, spark_key$Gene)]
# 
# # Aggregate by phylum
# spark_phyla <- aggregate(value ~ Taxonomy, data = snorm, mean)
# 
# # Keep the top half of both expressed and abundant phyla for plotting
# keep_phyla <- c(spark_metaG_phyla$Phylum[which(spark_metaG_phyla$value > quantile(spark_metaG_phyla$value)[3])], spark_phyla$Taxonomy[which(spark_phyla$value > quantile(spark_phyla$value)[3])])
# spark_phyla <- spark_phyla[which(spark_phyla$Taxonomy %in% keep_phyla),]
# spark_phyla$metaG <- spark_metaG_phyla$value[match(spark_phyla$Taxonomy, spark_metaG_phyla$Phylum)]
# spark_phyla$type <- type_key$type[match(spark_phyla$Taxonomy, type_key$phylum)]
# spark_phyla <- spark_phyla[which(is.na(spark_phyla$metaG) == F),]
# 
# p2 <- ggplot(spark_phyla, aes(x = metaG, y = value, color = type)) + geom_point(size = 2.5) + geom_label_repel(aes(label = Taxonomy, color = type), force = 5, size = 7.5) + scale_color_manual(values = c("royalblue", "goldenrod")) + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean Transcripts/L assigned", title = "Sparkling Lake") + theme(legend.position = "none")
# 
# #Process Trout expression data
# 
# # Similar to above, add phylum column and change weird formats
trout_key$Taxonomy <- gsub("Bacteria;", "", trout_key$Taxonomy)
trout_key$Taxonomy <- gsub("Proteobacteria;", "", trout_key$Taxonomy)
trout_key$Taxonomy <- gsub("Eukaryota;", "", trout_key$Taxonomy)
trout_key$Phylum <- sapply(strsplit(as.character(trout_key$Taxonomy),";"), `[`, 1)

trout_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("None", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified unclassified", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("Unclassified ", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", trout_key$Phylum)
trout_key$Phylum[grep("Blank", trout_key$Phylum)] <- "Unclassified"

# # Remove unclassified genes and the internal standard
# trout_key <- trout_key[grep("Unclassified|internal standard", trout_key$Phylum, invert = T), ]
# 
# # Prep read count table and switch to long format
# tnorm$Genes <- rownames(tnorm)
# tnorm <- tnorm[which(tnorm$Genes %in% trout_key$Gene), ]
# colnames(tnorm) <- gsub(".nonrRNA", "", colnames(tnorm))
# tnorm <- melt(tnorm)
# # Add columns of timepoint and phylum
# tnorm$Timepoint <- metadata$Timepoint[match(tnorm$variable, metadata$Sample)]
# tnorm$Taxonomy <- trout_key$Phylum[match(tnorm$Genes, trout_key$Gene)]
# 
# # Aggregate by phylum
# trout_phyla <- aggregate(value ~ Taxonomy, data = tnorm, mean)
# 
# # Keep the top half of both expressed and abundant phyla for plotting
# keep_phyla <- c(trout_metaG_phyla$Phylum[which(trout_metaG_phyla$value > quantile(trout_metaG_phyla$value)[3])], trout_phyla$Taxonomy[which(trout_phyla$value > quantile(trout_phyla$value)[3])])
# trout_phyla <- trout_phyla[which(trout_phyla$Taxonomy %in% keep_phyla),]
# trout_phyla$metaG <- trout_metaG_phyla$value[match(trout_phyla$Taxonomy, trout_metaG_phyla$Phylum)]
# trout_phyla$type <- type_key$type[match(trout_phyla$Taxonomy, type_key$phylum)]
# trout_phyla <- trout_phyla[which(is.na(trout_phyla$metaG) == F),]
# 
# p3 <- ggplot(trout_phyla, aes(x = metaG, y = value, color = type)) + geom_point(size = 2.5) + geom_label_repel(aes(label = Taxonomy, color = type), force = 5, size = 7.5) + scale_color_manual(values = c("royalblue", "goldenrod")) + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean Transcripts/L assigned", title = "Trout Bog") + theme(legend.position = "none")
# 
# plot_grid(p1, p2, p3, nrow = 1)
```

### The figure 1 chunk is hefty and needs to be run on my work desktop

**Figure 1. Most highly expressed or abundant phyla by lake.** To determine which phyla were most abundant or most expressed during our time series, we analyzed metagenomic and metatranscriptomic read counts. The expression of clustered, nonredundant genes was aggregated by phylum and compared to the coverage of those phyla in metagenomes. Genes that could not be classified in a phylum were not included in this analysis. No positive relationship was observed between expression and abundance. The identity of the most expressed and most abundant phyla varied by lake. One phylum, Chloroflexi, was removed from the plot of Lake Mendota due to orders of magnitude higher expression and abundance. This phylum is likely an outlier.

```{r, eval = F, echo = F}
# Write a function to calculate CoV within replicates and average by gene
within_reps <- function(readcounts){
  readcounts2 <- readcounts[which(rowSums(readcounts) > 1000), ]
  colnames(readcounts2) <- gsub(".nonrRNA", "", colnames(readcounts2))
  timepoints <- metadata$Timepoint[match(colnames(readcounts2), metadata$Sample)]
  timelist <- unique(timepoints)
  time_table <- readcounts2[,which(timepoints == timelist[1])]
  results <- data.frame(timepoint1 = apply(time_table, 1, cv))
  
  for(i in 2:length(timelist)){
    if(length(which(timepoints == timelist[i])) > 1){
      time_table <- readcounts2[,which(timepoints == timelist[i])]
      next_timepoint <- data.frame(apply(time_table, 1, cv))
      colnames(next_timepoint) <- paste("timepoint", i, sep = "")
      results <- cbind(results, next_timepoint)
    }
  }
  return(results)
}

mendota_within <- within_reps(mnorm)
mendota_within <- apply(mendota_within, 1, mean, na.rm = T)
mendota_across <- apply(mnorm, 1, cv, na.rm = T)
cv_stat <- c(rep("within replicates", length(mendota_within)), rep("across samples", length(mendota_across)))
mendota_cv_stats <- data.frame(type = cv_stat, stat = c(mendota_within, mendota_across))

p1 <- ggplot(mendota_cv_stats, aes(stat, fill = type)) + geom_density(alpha = 0.5) + scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + labs(x = "Coefficent of Variance (%)", title = "Lake Mendota") + theme(legend.title = element_blank())

spark_within <- within_reps(snorm)
spark_within <- apply(spark_within, 1, mean, na.rm = T)
spark_across <- apply(snorm, 1, cv, na.rm = T)
cv_stat <- c(rep("within replicates", length(spark_within)), rep("across samples", length(spark_across)))
spark_cv_stats <- data.frame(type = cv_stat, stat = c(spark_within, spark_across))

p2 <- ggplot(spark_cv_stats, aes(stat, fill = type)) + geom_density(alpha = 0.5) + scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + labs(x = "Coefficent of Variance (%)", title = "Sparkling Lake") + theme(legend.title = element_blank())


trout_within <- within_reps(tnorm)
trout_within <- apply(trout_within, 1, mean, na.rm = T)
trout_across <- apply(tnorm, 1, cv, na.rm = T)
cv_stat <- c(rep("within replicates", length(trout_within)), rep("across samples", length(trout_across)))
trout_cv_stats <- data.frame(type = cv_stat, stat = c(trout_within, trout_across))

p3 <- ggplot(trout_cv_stats, aes(stat, fill = type)) + geom_density(alpha = 0.5) + scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + labs(x = "Coefficent of Variance (%)", title = "Trout Bog") + theme(legend.title = element_blank())

plot_grid(p1, p2, p3, nrow = 3, labels = c("A", "B", "C"))
```

### Revise to include 1 lake since CoV is pretty equivalent between them. (so 4 panels total). Move the rest to supplemental

**Figure 2. Assessing the variability of metatranscriptomic read counts.** One aim of this metatranscriptomic study was to provide information on the variability in gene expression in freshwater that could be used to guide further metatranscriptomic experiments. We calculated the coefficient of variance (CoV) for each gene both within replicates (A) and across replicates (B). High levels of variability were observed in genes from all three lakes. In panels C-E, example trends over time of genes are shown. 


```{r, eval = T, warning = F, message = F, echo = F}

##### DESeq differential expression testing

abun_mnorm <- mnorm[order(rowSums(mnorm), decreasing = T), ]
abun_mnorm <- abun_mnorm[1:20000,]

colnames(abun_mnorm) <- gsub(".nonrRNA", "", colnames(abun_mnorm))

input <- as.matrix(abun_mnorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_mnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_mnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)
sig.res <- res[which(res$padj < 0.05), ]
sig.res.day <- sig.res[which(sig.res$log2FoldChange < 0), ]
sig.res.night <- sig.res[which(sig.res$log2FoldChange > 0), ]
sig.res.day.key <- mendota_key[match(rownames(sig.res.day), mendota_key$Gene),]
sig.res.night.key <- mendota_key[match(rownames(sig.res.night), mendota_key$Gene),]

# Make plots of this aggreagted by short taxa and groupd products

sig.res.day.key$Taxonomy <- gsub(";;;", "", sig.res.day.key$Taxonomy)
sig.res.day.key$Taxonomy <- gsub(";;", "", sig.res.day.key$Taxonomy)
sig.res.day.key$Taxonomy <- gsub(";$", "", sig.res.day.key$Taxonomy)
spl <- strsplit(as.character(sig.res.day.key$Taxonomy), ";")
sig.res.day.key$ShortTax <- sapply(lapply(spl, tail, 1), paste, collapse=";")

sig.res.night.key$Taxonomy <- gsub(";;;", "", sig.res.night.key$Taxonomy)
sig.res.night.key$Taxonomy <- gsub(";;", "", sig.res.night.key$Taxonomy)
sig.res.night.key$Taxonomy <- gsub(";$", "", sig.res.night.key$Taxonomy)
spl <- strsplit(as.character(sig.res.night.key$Taxonomy), ";")
sig.res.night.key$ShortTax <- sapply(lapply(spl, tail, 1), paste, collapse=";")

# Make column by certain key words 
sig.res.day.key$Category <- "None"
sig.res.day.key$Category[grep("photo|Photo", sig.res.day.key$Product)] <- "Photosynthesis"
sig.res.day.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.day.key$Product)] <- "Rhodopsin"
sig.res.day.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.day.key$Product)] <- "Sugar degradation"
sig.res.day.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.day.key$Product)] <- "RuBisCO"
sig.res.day.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.day.key$Product)] <- "Polyamines"
sig.res.day.key$Category[grep("citrate lyase|Citrate lyase", sig.res.day.key$Product)] <- "rTCA"
sig.res.day.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.day.key$Product)] <- "Nitrogenase"
sig.res.day.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.day.key$Product)] <- "Chitinase"
sig.res.day.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.day.key$Product)] <- "Glycoside_Hydrolase"
sig.res.day.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.day.key$Product)] <- "Alkaline_phosphatase"
sig.res.day.key$Category[grep("cellulase|cellulose", sig.res.day.key$Product)] <- "Cellulase"
sig.res.day.key$Category[grep("peroxidase|peroxide|catalase", sig.res.day.key$Product)] <- "ROS"
sig.res.day.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.day.key$Product)] <- "Methane/Ammonia"
sig.res.day.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.day.key$Product)] <- "Nitrite_reduction"
sig.res.day.key$Category[grep("urease", sig.res.day.key$Product)] <- "Urease"
sig.res.day.key$Category[grep("protease", sig.res.day.key$Product)] <- "Protease"

sig.res.night.key$Category <- "None"
sig.res.night.key$Category[grep("photo|Photo", sig.res.night.key$Product)] <- "Photosynthesis"
sig.res.night.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.night.key$Product)] <- "Rhodopsin"
sig.res.night.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.night.key$Product)] <- "Sugar degradation"
sig.res.night.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.night.key$Product)] <- "RuBisCO"
sig.res.night.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.night.key$Product)] <- "Polyamines"
sig.res.night.key$Category[grep("citrate lyase|Citrate lyase", sig.res.night.key$Product)] <- "rTCA"
sig.res.night.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.night.key$Product)] <- "Nitrogenase"
sig.res.night.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.night.key$Product)] <- "Chitinase"
sig.res.night.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.night.key$Product)] <- "Glycoside_Hydrolase"
sig.res.night.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.night.key$Product)] <- "Alkaline_phosphatase"
sig.res.night.key$Category[grep("cellulase|cellulose", sig.res.night.key$Product)] <- "Cellulase"
sig.res.night.key$Category[grep("peroxidase|peroxide|catalase", sig.res.night.key$Product)] <- "ROS"
sig.res.night.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.night.key$Product)] <- "Methane/Ammonia"
sig.res.night.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.night.key$Product)] <- "Nitrite_reduction"
sig.res.night.key$Category[grep("urease", sig.res.night.key$Product)] <- "Urease"
sig.res.night.key$Category[grep("protease", sig.res.night.key$Product)] <- "Protease"

sig.res.day.key$Condition <- "day"
sig.res.night.key$Condition <- "night"

sig.abun_mnorm_day <- abun_mnorm[match(sig.res.day.key$Gene, rownames(abun_mnorm)), which(conditions == "day")]
sig.res.day.key$totals <- rowSums(sig.abun_mnorm_day)
sig.abun_mnorm_night <- abun_mnorm[match(sig.res.night.key$Gene, rownames(abun_mnorm)), which(conditions == "night")]
sig.res.night.key$totals <- rowSums(sig.abun_mnorm_night)

sig.res.key <- rbind(sig.res.day.key, sig.res.night.key)
# Add in zero counts so bars are the same widths

dummy <- data.frame("Gene" = rep("nogene", 4), "Genome" = rep("nogenome", 4), "Taxonomy" = rep("notax", 4), "Product" = rep("noproduct", 4), "Phylum" = rep("Unclassified", 4), "ShortTax" = rep("noshorttax", 4), "Category" = c("Alkaline_phosphatase", "Chitinase", "Nitrite_reduction", "Polyamines"), "Condition" = c("night", "day", "night", "day"), "totals" = c(0, 0, 0, 0))

sig.res.key <- rbind(sig.res.key, dummy)
#sig.res.key$Phylum <- mendota_key$Phylum[match(sig.res.key$Gene, mendota_key$Gene)]

simple.sig.res.key <- aggregate(totals ~ Category + Condition, sig.res.key, sum)
simple.sig.res.key$totals[which(simple.sig.res.key$totals > 150000000000)] <- 150000000000
simple.sig.res.key <- simple.sig.res.key[grep("Photosynthesis|Polyamines|Rhodopsin|ROS|RuBisCO|Sugar degradation", simple.sig.res.key$Category), ]

p1 <- ggplot(data = simple.sig.res.key[which(simple.sig.res.key$Category != "None"), ], aes(x = Category, y = totals, fill = Condition)) + geom_bar(stat = "identity", position = "dodge") + labs(title = "Mendota", x = NULL, y = "Transcripts/L") + scale_fill_manual(values = c("darkgoldenrod2", "royalblue4")) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0))

# Plot taxonomy information as proportional bars
sig.res.key2 <- sig.res.key[grep("Photosynthesis|Polyamines|Rhodopsin|ROS|RuBisCO|Sugar degradation", sig.res.key$Category), ]
sig.res.key2$Phylum[which(is.na(sig.res.key2$Phylum) == T)] <- "Unclassified"

p2 <- ggplot(data = sig.res.key2[which(sig.res.key2$Condition == "day"), ], aes(x = Category, y = totals, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Day")
#+ scale_fill_manual(values = c("#ff7f00", "#1f78b4", "#33a02c", "#e31a1c", "#ffff99", "darkgrey", "#6a3d9a"))

p3 <- ggplot(data = sig.res.key2[which(sig.res.key2$Condition == "night"), ], aes(x = Category, y = totals, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Night")
#+ scale_fill_manual(values = c("#ff7f00", "#1f78b4", "#b2df8a", "#33a02c", "#b15928", "#ffff99", "darkgrey", "#6a3d9a"))

#plot_grid(p1, p2, p3, nrow = 1, labels = c("A", "B", "C"))
p1
p2
p3

# Repeat with Sparkling

abun_snorm <- snorm[order(rowSums(snorm), decreasing = T), ]
abun_snorm <- abun_snorm[1:20000,]

colnames(abun_snorm) <- gsub(".nonrRNA", "", colnames(abun_snorm))

input <- as.matrix(abun_snorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_snorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_snorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)
sig.res <- res[which(res$padj < 0.05), ]
sig.res.day <- sig.res[which(sig.res$log2FoldChange < 0), ]
sig.res.night <- sig.res[which(sig.res$log2FoldChange > 0), ]
sig.res.day.key <- spark_key[match(rownames(sig.res.day), spark_key$Gene),]
sig.res.night.key <- spark_key[match(rownames(sig.res.night), spark_key$Gene),]

# Make plots of this aggreagted by short taxa and groupd products

sig.res.day.key$Taxonomy <- gsub(";;;", "", sig.res.day.key$Taxonomy)
sig.res.day.key$Taxonomy <- gsub(";;", "", sig.res.day.key$Taxonomy)
sig.res.day.key$Taxonomy <- gsub(";$", "", sig.res.day.key$Taxonomy)
spl <- strsplit(as.character(sig.res.day.key$Taxonomy), ";")
sig.res.day.key$ShortTax <- sapply(lapply(spl, tail, 1), paste, collapse=";")

sig.res.night.key$Taxonomy <- gsub(";;;", "", sig.res.night.key$Taxonomy)
sig.res.night.key$Taxonomy <- gsub(";;", "", sig.res.night.key$Taxonomy)
sig.res.night.key$Taxonomy <- gsub(";$", "", sig.res.night.key$Taxonomy)
spl <- strsplit(as.character(sig.res.night.key$Taxonomy), ";")
sig.res.night.key$ShortTax <- sapply(lapply(spl, tail, 1), paste, collapse=";")

# Make column by certain key words 
sig.res.day.key$Category <- "None"
sig.res.day.key$Category[grep("photo|Photo", sig.res.day.key$Product)] <- "Photosynthesis"
sig.res.day.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.day.key$Product)] <- "Rhodopsin"
sig.res.day.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.day.key$Product)] <- "Sugar degradation"
sig.res.day.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.day.key$Product)] <- "RuBisCO"
sig.res.day.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.day.key$Product)] <- "Polyamines"
sig.res.day.key$Category[grep("citrate lyase|Citrate lyase", sig.res.day.key$Product)] <- "rTCA"
sig.res.day.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.day.key$Product)] <- "Nitrogenase"
sig.res.day.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.day.key$Product)] <- "Chitinase"
sig.res.day.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.day.key$Product)] <- "Glycoside_Hydrolase"
sig.res.day.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.day.key$Product)] <- "Alkaline_phosphatase"
sig.res.day.key$Category[grep("cellulase|cellulose", sig.res.day.key$Product)] <- "Cellulase"
sig.res.day.key$Category[grep("peroxidase|peroxide|catalase", sig.res.day.key$Product)] <- "ROS"
sig.res.day.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.day.key$Product)] <- "Methane/Ammonia"
sig.res.day.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.day.key$Product)] <- "Nitrite_reduction"
sig.res.day.key$Category[grep("urease", sig.res.day.key$Product)] <- "Urease"
sig.res.day.key$Category[grep("protease", sig.res.day.key$Product)] <- "Protease"

sig.res.night.key$Category <- "None"
sig.res.night.key$Category[grep("photo|Photo", sig.res.night.key$Product)] <- "Photosynthesis"
sig.res.night.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.night.key$Product)] <- "Rhodopsin"
sig.res.night.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.night.key$Product)] <- "Sugar degradation"
sig.res.night.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.night.key$Product)] <- "RuBisCO"
sig.res.night.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.night.key$Product)] <- "Polyamines"
sig.res.night.key$Category[grep("citrate lyase|Citrate lyase", sig.res.night.key$Product)] <- "rTCA"
sig.res.night.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.night.key$Product)] <- "Nitrogenase"
sig.res.night.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.night.key$Product)] <- "Chitinase"
sig.res.night.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.night.key$Product)] <- "Glycoside_Hydrolase"
sig.res.night.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.night.key$Product)] <- "Alkaline_phosphatase"
sig.res.night.key$Category[grep("cellulase|cellulose", sig.res.night.key$Product)] <- "Cellulase"
sig.res.night.key$Category[grep("peroxidase|peroxide|catalase", sig.res.night.key$Product)] <- "ROS"
sig.res.night.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.night.key$Product)] <- "Methane/Ammonia"
sig.res.night.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.night.key$Product)] <- "Nitrite_reduction"
sig.res.night.key$Category[grep("urease", sig.res.night.key$Product)] <- "Urease"
sig.res.night.key$Category[grep("protease", sig.res.night.key$Product)] <- "Protease"

sig.res.day.key$Condition <- "day"
sig.res.night.key$Condition <- "night"

sig.abun_snorm_day <- abun_snorm[match(sig.res.day.key$Gene, rownames(abun_snorm)), which(conditions == "day")]
sig.res.day.key$totals <- rowSums(sig.abun_snorm_day)
sig.abun_snorm_night <- abun_snorm[match(sig.res.night.key$Gene, rownames(abun_snorm)), which(conditions == "night")]
sig.res.night.key$totals <- rowSums(sig.abun_snorm_night)

sig.res.key <- rbind(sig.res.day.key, sig.res.night.key)
# Add in zero counts so bars are the same widths

dummy <- data.frame("Gene" = rep("nogene", 4), "Genome" = rep("nogenome", 4), "Taxonomy" = rep("notax", 4), "Product" = rep("noproduct", 4), "Phylum" = rep("Unclassified", 4), "ShortTax" = rep("noshorttax", 4), "Category" = c("Alkaline_phosphatase", "Chitinase", "Nitrite_reduction", "Polyamines"), "Condition" = c("night", "day", "night", "day"), "totals" = c(0, 0, 0, 0))

sig.res.key <- rbind(sig.res.key, dummy)
#sig.res.key$Phylum <- spark_key$Phylum[match(sig.res.key$Gene, spark_key$Gene)]

simple.sig.res.key <- aggregate(totals ~ Category + Condition, sig.res.key, sum)
simple.sig.res.key$totals[which(simple.sig.res.key$totals > 25000000000)] <- 25000000000
simple.sig.res.key <- simple.sig.res.key[grep("Photosynthesis|Polyamines|Rhodopsin|ROS|RuBisCO|Sugar degradation", simple.sig.res.key$Category), ]

p1 <- ggplot(data = simple.sig.res.key[which(simple.sig.res.key$Category != "None"), ], aes(x = Category, y = totals, fill = Condition)) + geom_bar(stat = "identity", position = "dodge") + labs(title = "Sparkling Lake", x = NULL, y = "Transcripts/L") + scale_fill_manual(values = c("darkgoldenrod2", "royalblue4")) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0))

# Plot taxonomy information as proportional bars
sig.res.key2 <- sig.res.key[grep("Photosynthesis|Polyamines|Rhodopsin|ROS|RuBisCO|Sugar degradation", sig.res.key$Category), ]
sig.res.key2$Phylum[which(is.na(sig.res.key2$Phylum) == T)] <- "Unclassified"

p2 <- ggplot(data = sig.res.key2[which(sig.res.key2$Condition == "day"), ], aes(x = Category, y = totals, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Day")
#+ scale_fill_manual(values = c("#ff7f00", "#1f78b4", "#33a02c", "#e31a1c", "#ffff99", "darkgrey", "#6a3d9a"))

p3 <- ggplot(data = sig.res.key2[which(sig.res.key2$Condition == "night"), ], aes(x = Category, y = totals, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Night")
#+ scale_fill_manual(values = c("#ff7f00", "#1f78b4", "#b2df8a", "#33a02c", "#b15928", "#ffff99", "darkgrey", "#6a3d9a"))

#plot_grid(p1, p2, p3, nrow = 1, labels = c("A", "B", "C"))
p1
p2
p3

# Repeat with Trout Bog

abun_tnorm <- tnorm[order(rowSums(tnorm), decreasing = T), ]
abun_tnorm <- abun_tnorm[1:20000,]

colnames(abun_tnorm) <- gsub(".nonrRNA", "", colnames(abun_tnorm))

input <- as.matrix(abun_tnorm)
input <- input/500
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_tnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_tnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)
sig.res <- res[which(res$padj < 0.05), ]
sig.res.day <- sig.res[which(sig.res$log2FoldChange < 0), ]
sig.res.night <- sig.res[which(sig.res$log2FoldChange > 0), ]
sig.res.day.key <- trout_key[match(rownames(sig.res.day), trout_key$Gene),]
sig.res.night.key <- trout_key[match(rownames(sig.res.night), trout_key$Gene),]

# Make plots of this aggreagted by short taxa and groupd products

sig.res.day.key$Taxonomy <- gsub(";;;", "", sig.res.day.key$Taxonomy)
sig.res.day.key$Taxonomy <- gsub(";;", "", sig.res.day.key$Taxonomy)
sig.res.day.key$Taxonomy <- gsub(";$", "", sig.res.day.key$Taxonomy)
spl <- strsplit(as.character(sig.res.day.key$Taxonomy), ";")
sig.res.day.key$ShortTax <- sapply(lapply(spl, tail, 1), paste, collapse=";")

sig.res.night.key$Taxonomy <- gsub(";;;", "", sig.res.night.key$Taxonomy)
sig.res.night.key$Taxonomy <- gsub(";;", "", sig.res.night.key$Taxonomy)
sig.res.night.key$Taxonomy <- gsub(";$", "", sig.res.night.key$Taxonomy)
spl <- strsplit(as.character(sig.res.night.key$Taxonomy), ";")
sig.res.night.key$ShortTax <- sapply(lapply(spl, tail, 1), paste, collapse=";")

# Make column by certain key words 
sig.res.day.key$Category <- "None"
sig.res.day.key$Category[grep("photo|Photo", sig.res.day.key$Product)] <- "Photosynthesis"
sig.res.day.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.day.key$Product)] <- "Rhodopsin"
sig.res.day.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.day.key$Product)] <- "Sugar degradation"
sig.res.day.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.day.key$Product)] <- "RuBisCO"
sig.res.day.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.day.key$Product)] <- "Polyamines"
sig.res.day.key$Category[grep("citrate lyase|Citrate lyase", sig.res.day.key$Product)] <- "rTCA"
sig.res.day.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.day.key$Product)] <- "Nitrogenase"
sig.res.day.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.day.key$Product)] <- "Chitinase"
sig.res.day.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.day.key$Product)] <- "Glycoside_Hydrolase"
sig.res.day.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.day.key$Product)] <- "Alkaline_phosphatase"
sig.res.day.key$Category[grep("cellulase|cellulose", sig.res.day.key$Product)] <- "Cellulase"
sig.res.day.key$Category[grep("peroxidase|peroxide|catalase", sig.res.day.key$Product)] <- "ROS"
sig.res.day.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.day.key$Product)] <- "Methane/Ammonia"
sig.res.day.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.day.key$Product)] <- "Nitrite_reduction"
sig.res.day.key$Category[grep("urease", sig.res.day.key$Product)] <- "Urease"
sig.res.day.key$Category[grep("protease", sig.res.day.key$Product)] <- "Protease"

sig.res.night.key$Category <- "None"
sig.res.night.key$Category[grep("photo|Photo", sig.res.night.key$Product)] <- "Photosynthesis"
sig.res.night.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.night.key$Product)] <- "Rhodopsin"
sig.res.night.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.night.key$Product)] <- "Sugar degradation"
sig.res.night.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.night.key$Product)] <- "RuBisCO"
sig.res.night.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.night.key$Product)] <- "Polyamines"
sig.res.night.key$Category[grep("citrate lyase|Citrate lyase", sig.res.night.key$Product)] <- "rTCA"
sig.res.night.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.night.key$Product)] <- "Nitrogenase"
sig.res.night.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.night.key$Product)] <- "Chitinase"
sig.res.night.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.night.key$Product)] <- "Glycoside_Hydrolase"
sig.res.night.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.night.key$Product)] <- "Alkaline_phosphatase"
sig.res.night.key$Category[grep("cellulase|cellulose", sig.res.night.key$Product)] <- "Cellulase"
sig.res.night.key$Category[grep("peroxidase|peroxide|catalase", sig.res.night.key$Product)] <- "ROS"
sig.res.night.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.night.key$Product)] <- "Methane/Ammonia"
sig.res.night.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.night.key$Product)] <- "Nitrite_reduction"
sig.res.night.key$Category[grep("urease", sig.res.night.key$Product)] <- "Urease"
sig.res.night.key$Category[grep("protease", sig.res.night.key$Product)] <- "Protease"

sig.res.day.key$Condition <- "day"
sig.res.night.key$Condition <- "night"

sig.abun_tnorm_day <- abun_tnorm[match(sig.res.day.key$Gene, rownames(abun_tnorm)), which(conditions == "day")]
sig.res.day.key$totals <- rowSums(sig.abun_tnorm_day)
sig.abun_tnorm_night <- abun_tnorm[match(sig.res.night.key$Gene, rownames(abun_tnorm)), which(conditions == "night")]
sig.res.night.key$totals <- rowSums(sig.abun_tnorm_night)

sig.res.key <- rbind(sig.res.day.key, sig.res.night.key)
# Add in zero counts so bars are the same widths

dummy <- data.frame("Gene" = rep("nogene", 4), "Genome" = rep("nogenome", 4), "Taxonomy" = rep("notax", 4), "Product" = rep("noproduct", 4), "Phylum" = rep("Unclassified", 4), "ShortTax" = rep("noshorttax", 4), "Category" = c("Alkaline_phosphatase", "Chitinase", "Nitrite_reduction", "Polyamines"), "Condition" = c("night", "day", "night", "day"), "totals" = c(0, 0, 0, 0))

sig.res.key <- rbind(sig.res.key, dummy)
#sig.res.key$Phylum <- trout_key$Phylum[match(sig.res.key$Gene, trout_key$Gene)]

simple.sig.res.key <- aggregate(totals ~ Category + Condition, sig.res.key, sum)
simple.sig.res.key$totals[which(simple.sig.res.key$totals > 500000000000)] <- 500000000000
simple.sig.res.key <- simple.sig.res.key[grep("Photosynthesis|Polyamines|Rhodopsin|ROS|RuBisCO|Sugar degradation", simple.sig.res.key$Category), ]

p1 <- ggplot(data = simple.sig.res.key[which(simple.sig.res.key$Category != "None"), ], aes(x = Category, y = totals, fill = Condition)) + geom_bar(stat = "identity", position = "dodge") + labs(title = "Trout Bog", x = NULL, y = "Transcripts/L") + scale_fill_manual(values = c("darkgoldenrod2", "royalblue4")) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0))

# Plot taxonomy information as proportional bars
sig.res.key2 <- sig.res.key[grep("Photosynthesis|Polyamines|Rhodopsin|ROS|RuBisCO|Sugar degradation", sig.res.key$Category), ]
sig.res.key2$Phylum[which(is.na(sig.res.key2$Phylum) == T)] <- "Unclassified"

p2 <- ggplot(data = sig.res.key2[which(sig.res.key2$Condition == "day"), ], aes(x = Category, y = totals, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Day")
#+ scale_fill_manual(values = c("#ff7f00", "#1f78b4", "#33a02c", "#e31a1c", "#ffff99", "darkgrey", "#6a3d9a"))

p3 <- ggplot(data = sig.res.key2[which(sig.res.key2$Condition == "night"), ], aes(x = Category, y = totals, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Night")
#+ scale_fill_manual(values = c("#ff7f00", "#1f78b4", "#b2df8a", "#33a02c", "#b15928", "#ffff99", "darkgrey", "#6a3d9a"))

#plot_grid(p1, p2, p3, nrow = 1, labels = c("A", "B", "C"))
p1
p2
p3
```

# include carboxylates. split proteos. suppress errors. repeat by lake. find way to make phylum legend universal? use zero entries?


**Figure 3. Differential expression in day vs. night in Lake Mendota.** Many genes were differentially expressed depending on light conditions in Lake Mendota. For example, sugar transport was more highly expressed at night, while nitrogen cycling genes were more highly expressed during the day. 