---
author: "Alex Linz"
date: "August 7, 2018"
---
<style>
body {
text-align: justify}
</style>


# Metatranscriptomics reveals interactions between phototrophs and heterotrophs in freshwater

```{r, Setup, eval = T, echo = F, message = F, warning = F}
# Set up environment
path <- "D:/"
path2 <- "C:/Users/Alex/"

library(ggplot2)
library(cowplot)
library(reshape2)
library(ggrepel)
# library(DESeq2)

zscore <- function(counts){
  z <- (counts - mean(counts)) / sd(counts)
  return(z)
}

# Metatranscriptome data
snorm <- read.csv(paste(path, "geodes_data_tables/Sparkling_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
tnorm <- read.csv(paste(path, "geodes_data_tables/Trout_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
mnorm <- read.csv(paste(path, "geodes_data_tables/Mendota_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)

snorm <- snorm[, which(colnames(snorm) != "GEODES014.nonrRNA" & colnames(snorm) != "GEODES033.nonrRNA")]
tnorm <- tnorm[, which(colnames(tnorm) != "GEODES065.nonrRNA")]
mnorm <- mnorm[, which(colnames(mnorm) != "GEODES158.nonrRNA")]

mendota_key <- read.csv(paste(path, "geodes_data_tables/Mendota_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)
spark_key <- read.csv(paste(path, "geodes_data_tables/Sparkling_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)
trout_key <- read.csv(paste(path, "geodes_data_tables/Trout_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)

# Sample data
metadata <- read.csv(file = paste(path2, "Desktop/geodes/bioinformatics_workflow/R_processing/sample_metadata.csv", sep = ""), header = T)
enviro_data <- read.csv(paste(path2, "Desktop/geodes/environmental_data/for_humans/compiled_field_data_for_R.csv", sep = ""), header = T)
enviro_data_no_sonde <- enviro_data[which(is.na(enviro_data$pH) == F), ]
chemdata <- read.csv(paste(path2, "Desktop/geodes/environmental_data/GEODES_TNTP.csv", sep = ""), header = T)
chemdata$Lake <- NA
chemdata$Lake[grep("ME", chemdata$Sample)] <- "Mendota"
chemdata$Lake[grep("SP", chemdata$Sample)] <- "Sparkling"
chemdata$Lake[grep("TB", chemdata$Sample)] <- "Trout"

# Metagenome data
metaG_reads <- read.table(paste(path, "geodes_data_tables/GEODES_metaG_ID90_2018-03-10.readcounts.txt", sep = ""), row.names = 1, sep = "\t")
colnames(metaG_reads) <- c("GEODES005", "GEODES006", "GEODES057", "GEODES058", "GEODES117", "GEODES118", "GEODES165", "GEODES166", "GEODES167", "GEODES168")
metaG_key <- read.table(paste(path, "geodes_data_tables/GEODES_metaG_genekey_2018-03-12.txt", sep = ""), sep = "\t", quote = "")
colnames(metaG_key) <- c("Gene", "Genome", "Taxonomy", "Product")
lakekey <- c("Sparkling", "Sparkling", "Trout", "Trout", "Mendota", "Mendota", "Sparkling2009", "Sparkling2009", "Sparkling2009", "Sparkling2009")
metaG_reads <- sweep(metaG_reads, 2, colSums(metaG_reads), "/")

```

### Main text tables and figures.

**Table 1. Comparison of Sparkling Lake, Lake Mendota, and Trout Bog.** These three lakes were chosen for comparative metatranscriptomics because of their varying trophic statuses, extensive historical data, and previous microbial sampling. Data on surface area, maximum depth, and development on shoreline courtesy of NTL-LTER <lter.limnology.wisc.edu>. Temperature, dissolved oxygen, pH, and conductivity were measured using a HydroLab DS5x Sonde and are averaged over all sampling depths and timepoints for each lake. Chlorophyll and phycocyanin concentrations were measured from the integrated epilimnion samples using a methanol extraction protocol and averaged over all timepoints. Secchi depth was measured at the first timepoint for each lake. Bacterial production was quantified via C14-leucine incorporation and averaged over all timepoints. Due to thunderstorms the night of July 8th, the final 1AM timepoint in Sparkling Lake was collected on July 9th instead. 

|                         | Lake Mendota      | Trout Bog                | Sparkling Lake           |
| ----------------------- |:-----------------:|:------------------------:|:------------------------:|
| Surface area (km^2)     | 39.6              | 0.001                    | 0.637                    |
| Maximum depth (m)       | 25.3              | 7.9                      | 20                       |
| Trophic status          | Eutrophic         | Humic                    | Oligotrophic             |
| Location                | Madison, WI USA   | Boulder Junction, WI USA | Boulder Junction, WI USA |
| GPS Coordinates         | 43.1113, -89.4255 | 46.0412, -89.6861        | 46.0091, -89.6695        |
| Shoreline development   | High              | Low                      | Moderate                 |
| Epilimnion sampling depth| 0-7m             | 0-1.5m                   | 0-4m                     |
| Temperature (C)         | `r round(mean(enviro_data$Temperature[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$Temperature[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$Temperature[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
|Dissolved oxygen (mg/L)  | `r round(mean(enviro_data$DO[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$DO[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$DO[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| pH                      | `r round(mean(enviro_data$pH[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$pH[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$pH[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| Conductivity (uS/cm)    | `r round(mean(enviro_data$Conductivity[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$Conductivity[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$Conductivity[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| Total phosphorus (ug/L) | `r round(mean(chemdata$TP[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TP[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TP[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total nitrogen (ug/L)   | `r round(mean(chemdata$TN[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TN[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TN[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total dissolved phosphorus (ug/L) | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total dissolved nitrogen (ug/L)   | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Sparkling")]), 2)` |
| Chlorophyll (ug/L)      | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7 & enviro_data$Timepoint != 40 & enviro_data$Timepoint != 44)]), 2)` | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5 & enviro_data$Timepoint != 20)]), 2)` | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Phycocyanin (ug/L)      | `r round(mean(enviro_data$Phycocyanin[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7 & enviro_data$Timepoint != 40 & enviro_data$Timepoint != 44)]), 2)` | `r round(mean(enviro_data$Phycocyanin[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5 & enviro_data$Timepoint != 20)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Bacterial production (cpm)    | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Secchi depth (m)        | 4.8               | 1.1                      | 6.2                      |
| Sampling dates (2016)   | July 14-16        | July 8-10                | July 6-9                 |
| Sunrise/sunset time     | 5:32/20:35        | 5:18/20:49               | 5:17/20:50               |

```{r, table2, eval = F, echo = F}
totals <- rowSums(snorm)
top10 <- rownames(snorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
spark_key[match(top10, spark_key$Gene),]

totals <- rowSums(mnorm)
top10 <- rownames(mnorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
mendota_key[match(top10, mendota_key$Gene),]

totals <- rowSums(tnorm)
top10 <- rownames(tnorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
trout_key[match(top10, trout_key$Gene),]

```

**Table 2. Top 10 most expressed genes in each study site.** Coding regions from reference genomes and metagenome assemblies were clustered at 97% sequence similarity, and the longest coding region was chosen as the representative sequence. Metatranscriptomic reads were mapped to these representative sequences. The top 10 most expressed sequences from each lake, without any filters, are presented here. Annotations and classifications are derived from the sequence to which each read mapped.


| Lake Mendota      | Trout Bog                | Sparkling Lake           |
|:-----------------|:------------------------|:------------------------|
| Photosystem II P680 reaction center D1 protein (Cyanobacteria) | Photosystem II P680 reaction center D1 protein (unclassified) | Hypothetical protein (unclassified) |
| Photosystem II P680 reaction center D1 protein (unclassified) | Ribulose-bisphosphate carboxylase large chain (unclassified) | Hypothetical protein (unclassified) |
| Photosystem II P680 reaction center D1 protein (Cyanobacteria) | Photosystem II CP43 chlorophyll apoprotein (unclassified) | Photosystem II P680 reaction center D1 protein (Cyanobacteria) |
| Hypothetical protein (unclassified) | Putative beta-barrel porin2, OmpL-like (unclassified) | Photosystem II P680 reaction center D1 protein (unclassified) |
| Hypothetical protein (unclassified) | Photosystem II P680 reaction center D1 protein (Cyanobacteria) | Photosystem II P680 reaction center D1 protein (Chitinophagaceae) |
| Ribulose-bisphosphate carboxylase large chain (Cyanobacteria) | Photosystem II P700 chlorophyll a apoprotein  A2 (unclassified) | Hypothetical protein (unclassified) |
| PQQ-dependent dehydrogenase (LD28) | Photosystem II CP47 chlorophyll apoprotein (unclassified) | Hypothetical protein (unclassified) |
| Hypothetical protein (unclassified) | Hypothetical protein (unclassified) | Photosystem II P680 reaction center D1 protein (unclassified) |
| Hypothetical protein (Bdellovibrio) | Photosystem II P700 chlorophyll a apoprotein  A2 (unclassified) | Hypothetical protein (unclassified) |
| Ribulose-bisphosphate carboxylase large chain (Cyanobacteria) | Photosystem II P680 reaction center D1 protein (unclassified) | Hypothetical protein (unclassified) |


```{r, table3, eval = F, echo = F}
totals <- rowSums(snorm)
top10 <- rownames(snorm)[order(totals, decreasing = T)]
print <- spark_key[match(top10, spark_key$Gene),]
print <- print[grep("Cyano", print$Taxonomy, invert = T), ]
print <- print[grep("photo|Photo", print$Product, invert = T), ]
print <- print[grep("chloro|Chloro|ribulose-bisphosphate carboxylase", print$Product, invert = T), ]
print <- print[grep("hypothetical|Hypothetical|internal standard", print$Product, invert = T), ]
print[1:10,3:4]

totals <- rowSums(mnorm)
top10 <- rownames(mnorm)[order(totals, decreasing = T)]
print <- mendota_key[match(top10, mendota_key$Gene),]
print <- print[grep("Cyano", print$Taxonomy, invert = T), ]
print <- print[grep("photo|Photo", print$Product, invert = T), ]
print <- print[grep("chloro|Chloro|ribulose-bisphosphate carboxylase", print$Product, invert = T), ]
print <- print[grep("hypothetical|Hypothetical|internal standard", print$Product, invert = T), ]
print[1:10,3:4]

totals <- rowSums(tnorm)
top10 <- rownames(tnorm)[order(totals, decreasing = T)]
print <- trout_key[match(top10, trout_key$Gene),]
print <- print[grep("Cyano", print$Taxonomy, invert = T), ]
print <- print[grep("photo|Photo", print$Product, invert = T), ]
print <- print[grep("chloro|Chloro|ribulose-bisphosphate carboxylase", print$Product, invert = T), ]
print <- print[grep("hypothetical|Hypothetical|internal standard|unknown function", print$Product, invert = T), ]
print[1:10,3:4]

```
**Table 3. Top 10 most expressed annotated heterotrophic genes in each study site.** Coding regions from reference genomes and metagenome assemblies were clustered at 97% sequence similarity, and the longest coding region was chosen as the representative sequence. Metatranscriptomic reads were mapped to these representative sequences. The top 10 most expressed genes from each lake, filtered to exclude photosynthetic genes, phototrophic organisms, hypothetical genes, and unclassified genes, are presented here. Annotations and classifications are derived from the sequence to which each read mapped.

| Lake Mendota      | Trout Bog                | Sparkling Lake           |
|:-----------------|:------------------------|:------------------------|
| PQQ-dependent dehydrogenase (LD28) | Putative beta-barrel porin-2, OmpL-like (unclassified) | Chaperonin GroEL (Deltaproteobacteria) |
| Translation elongation factor TU (Bacteroidetes) | Ig-like domain group 3 (unclassified) | Fibronectin type III domain (unclassified) |
| YTV protein (Actinobacteria) | Ig-like domain group 1 (unclassified) | S-layer homology domain (Fimbriimonas ginsengisoli) |
| Translation elongation factor EF-G (acI-A6) | PEP-CTERM motif (unclassified) | Cytochrome c oxidase subunit 1 (unclassified) |
| DNA-directed RNA polymerase (acI-A6) | Cytochrome c oxidase subunit 1 (unclassified) | Cytochrome c oxidase subunit 2 (unclassified) |
| Probable sodium:solute symporter (LD12) | Cytochrome b6 (unclassified) | HYR domain, putative metal-binding (unclassified) |
| DNA-directed RNA polymerase (acI-A6) | Ig-like group 1 (acI-B) | F-type H+-transporting ATPase subunit beta (Chitinophagia) |
| Phycoerythrin alpha chain (unclassified) | S-layer homology domain (Fimbriimonas) | molecular chaperone HtpG (unclassified) |
| ABC-type sugar transport system (acI-B1) | Ribonucleoside-diphosphate reductase alpha chain (Pedosphaera parvula) | Glucose dehydrogenase (unclassified) |
| Translation elongation factor EF-G (acI-A6) | DNA-directed RNA polymerase subunit B (unclassified) | Lactocepin (unclassified) |



```{r, fig1, eval = T}
# Process metagenome gene key to include a phylum column. Fix any weird formats.

metaG_key$Taxonomy <- gsub("Bacteria;", "", metaG_key$Taxonomy)
metaG_key$Taxonomy <- gsub("Eukaryota;", "", metaG_key$Taxonomy)
metaG_key$Phylum <- sapply(strsplit(as.character(metaG_key$Taxonomy),";"), `[`, 1)

metaG_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("None", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Unclassified ", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("TM7", "Saccharibacteria", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Ignavibacteriae", "Ignavibacteria", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", metaG_key$Phylum)
metaG_key$Phylum[which(is.na(metaG_key$Phylum) == T)] <- "Unclassified"
metaG_key$Phylum[grep("Blank", metaG_key$Phylum)] <- "Unclassified"

# Remove unclassified genes to save on RAM
metaG_key <- metaG_key[which(metaG_key$Phylum != "Unclassified"),]

# Split metagenome read count tables by lake and prep for melting
metaG_reads$Genes <- rownames(metaG_reads)
metaG_reads <- metaG_reads[match(metaG_reads$Genes, metaG_key$Gene),]
spark_metaG <- metaG_reads[,c(1,2, 11)]
trout_metaG <- metaG_reads[,c(3,4, 11)]
mendota_metaG <- metaG_reads[,c(5,6, 11)]

# Use melt to switch from wide to long format
spark_metaG <- melt(spark_metaG)
trout_metaG <- melt(trout_metaG)
mendota_metaG <- melt(mendota_metaG)

# Add phylum column to each melted lake table
mendota_metaG$Phylum <- metaG_key$Phylum[match(mendota_metaG$Genes, metaG_key$Gene)]
trout_metaG$Phylum <- metaG_key$Phylum[match(trout_metaG$Genes, metaG_key$Gene)]
spark_metaG$Phylum <- metaG_key$Phylum[match(spark_metaG$Genes, metaG_key$Gene)]

# Aggregate counts by phylum
mendota_metaG_phyla <- aggregate(value ~ Phylum, data = mendota_metaG, mean)
spark_metaG_phyla <- aggregate(value ~ Phylum, data = spark_metaG, mean)
trout_metaG_phyla <- aggregate(value ~ Phylum, data = trout_metaG, mean)

# # Remove metaG datasets not needed any more to save RAM
# rm(metaG_reads)
# rm(mendota_metaG)
# rm(trout_metaG)
# rm(spark_metaG)
# rm(metaG_key)

# Include a key for colorcoding phyla by broader classification

type_key <- data.frame(phylum = c("Actinobacteria", "Bacteroidetes", "Chloroflexi", "Crenarchaeota", "Cryptophyta", "Cyanobacteria", "Gemmatimonadetes", "Heterokonta", "Ignavibacteria", "Planctomycetes", "Proteobacteria", "Verrucomicrobia", "Viruses"), type = c("Bacteria", "Bacteria", "Bacteria", "Archaea", "Algae", "Bacteria", "Bacteria", "Algae", "Bacteria", "Bacteria", "Bacteria", "Bacteria", "Viruses"))


#Process Mendota expression data

# Similar to above, add phylum column and change weird formats
mendota_key$Taxonomy <- gsub("Bacteria;", "", mendota_key$Taxonomy)
mendota_key$Taxonomy <- gsub("Eukaryota;", "", mendota_key$Taxonomy)
mendota_key$Phylum <- sapply(strsplit(as.character(mendota_key$Taxonomy),";"), `[`, 1)

mendota_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("None", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Unclassified ", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", mendota_key$Phylum)
mendota_key$Phylum[grep("Blank", mendota_key$Phylum)] <- "Unclassified"

# Remove unclassified genes and the internal standard
mendota_key <- mendota_key[grep("Unclassified|internal standard", mendota_key$Phylum, invert = T), ]

# Prep read count table and switch to long format
mnorm$Genes <- rownames(mnorm)
mnorm <- mnorm[which(mnorm$Genes %in% mendota_key$Gene), ]
colnames(mnorm) <- gsub(".nonrRNA", "", colnames(mnorm))
mnorm <- melt(mnorm)
# Add columns of timepoint and phylum
mnorm$Timepoint <- metadata$Timepoint[match(mnorm$variable, metadata$Sample)]
mnorm$Taxonomy <- mendota_key$Phylum[match(mnorm$Genes, mendota_key$Gene)]

# Aggregate by phylum
mendota_phyla <- aggregate(value ~ Taxonomy, data = mnorm, mean)

# Keep the top half of both expressed and abundant phyla for plotting
keep_phyla <- c(mendota_metaG_phyla$Phylum[which(mendota_metaG_phyla$value > quantile(mendota_metaG_phyla$value)[3])], mendota_phyla$Taxonomy[which(mendota_phyla$value > quantile(mendota_phyla$value)[3])])
mendota_phyla <- mendota_phyla[which(mendota_phyla$Taxonomy %in% keep_phyla),]
mendota_phyla$metaG <- mendota_metaG_phyla$value[match(mendota_phyla$Taxonomy, mendota_metaG_phyla$Phylum)]
mendota_phyla$type <- type_key$type[match(mendota_phyla$Taxonomy, type_key$phylum)]
mendota_phyla <- mendota_phyla[which(is.na(mendota_phyla$metaG) == F),]

# Color key: c("limegreen", "lavenderblush4", "royalblue", "goldenrod")) == algae, archaea, bacteria, viruses
ggplot(mendota_phyla, aes(x = metaG, y = value, color = type)) + geom_point(size = 2.5) + geom_label_repel(aes(label = Taxonomy, color = type), force = 5, size = 7.5) + scale_color_manual(values = c("royalblue", "goldenrod")) + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean Transcripts/L assigned", title = "Lake Mendota") + theme(legend.position = "none")

#Clear mendota data from workspace and start with sparkling

# rm(averaged_tax)
# rm(mendota_intersect)
# rm(wide_mnorm)
# rm(mendota_metaG_phyla)
# rm(mendota_phyla)




```


**Figure 1 Most highly expressed or abundant phyla by lake.** The expression of clustered, nonredundant genes was aggregated by phylum and compared to the coverage of those phyla in metagenomes. Genes that could not be classified in a phylum were not included in this analysis. No positive relationship was observed between expression and abundance. The identity of the most expressed and most abundant phyla varied by lake. One phylum, Chloroflexi, was removed from the plot of Lake Mendota due to orders of magnitude higher expression and abundance. This phylum is likely an outlier.

