---
title: "GEODES_Figures"
author: "Alex Linz"
date: "10/18/2018"
output:
  html_document: default
  pdf_document: default
---

# Metatranscriptomics reveals interactions between phototrophs and heterotrophs in freshwater

new plan:
table 1: metadata
table 2-4: categories by lake - last column getting cutoff?
Figure 1: cyclic trends example
Figure 2: barplots category by lake 

table 2-4 plan:
category| % expressed in day (genes/reads) | % expressed in night (genes/reads) | % cyclic

<style>
body {
text-align: justify}
</style>

<style type="text/css">
.main-container {
  max-width: 3000px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r, Setup, eval = T, echo = F, message = F, warning = F}
# Set up environment
path <- "/Users/Alex/Desktop/"
path2 <- "/Users/Alex/"

library(ggplot2)
library(cowplot)
library(reshape2)
library(ggrepel)
library(raster)
library(DESeq2)
library(rain)

zscore <- function(counts){
  z <- (counts - mean(counts)) / sd(counts)
  return(z)
}

# Sample data
metadata <- read.csv(file = paste(path2, "Desktop/geodes/bioinformatics_workflow/R_processing/sample_metadata.csv", sep = ""), header = T)
enviro_data <- read.csv(paste(path2, "Desktop/geodes/environmental_data/for_humans/compiled_field_data_for_R.csv", sep = ""), header = T)
enviro_data_no_sonde <- enviro_data[which(is.na(enviro_data$pH) == F), ]
chemdata <- read.csv(paste(path2, "Desktop/geodes/environmental_data/GEODES_TNTP.csv", sep = ""), header = T)
chemdata$Lake <- NA
chemdata$Lake[grep("ME", chemdata$Sample)] <- "Mendota"
chemdata$Lake[grep("SP", chemdata$Sample)] <- "Sparkling"
chemdata$Lake[grep("TB", chemdata$Sample)] <- "Trout"

```

```{r, Mendota, echo = F, eval = T}

mnorm <- read.csv(paste(path, "geodes_data_tables/Mendota_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
mnorm <- mnorm[, which(colnames(mnorm) != "GEODES158.nonrRNA")]
colnames(mnorm) <- gsub(".nonrRNA", "", colnames(mnorm))
mendota_key <- read.csv(paste(path, "geodes_data_tables/Mendota_ID90_genekey_geneclassifications_2018-11-28.csv", sep = ""), header = T)


```

```{r, Trout, echo = F, eval = T}

tnorm <- read.csv(paste(path, "geodes_data_tables/Trout_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
tnorm <- tnorm[, which(colnames(tnorm) != "GEODES065.nonrRNA")]
colnames(tnorm) <- gsub(".nonrRNA", "", colnames(tnorm))
trout_key <- read.csv(paste(path, "geodes_data_tables/Trout_ID90_genekey_geneclassifications_2018-11-28.csv", sep = ""), header = T)


```

```{r, Sparkling, echo = F, eval = T}
snorm <- read.csv(paste(path, "geodes_data_tables/Sparkling_ID90_normalized_readcounts.csv", sep = ""), header = T, row.names = 1)
snorm <- snorm[, which(colnames(snorm) != "GEODES014.nonrRNA" & colnames(snorm) != "GEODES033.nonrRNA")]
colnames(snorm) <- gsub(".nonrRNA", "", colnames(snorm))
spark_key <- read.csv(paste(path, "geodes_data_tables/Sparkling_ID90_genekey_geneclassifications_2018-11-28.csv", sep = ""), header = T)

```


```{r, Metagenomes, echo = F, eval = F}

# Metagenome data
metaG_reads <- read.table(paste(path, "geodes_data_tables/GEODES_metaG_ID90_2018-03-10.readcounts.txt", sep = ""), row.names = 1, sep = "\t")
colnames(metaG_reads) <- c("GEODES005", "GEODES006", "GEODES057", "GEODES058", "GEODES117", "GEODES118", "GEODES165", "GEODES166", "GEODES167", "GEODES168")
metaG_key <- read.table(paste(path, "geodes_data_tables/GEODES_metaG_genekey_2018-03-12.txt", sep = ""), sep = "\t", quote = "")
colnames(metaG_key) <- c("Gene", "Genome", "Taxonomy", "Product")
lakekey <- c("Sparkling", "Sparkling", "Trout", "Trout", "Mendota", "Mendota", "Sparkling2009", "Sparkling2009", "Sparkling2009", "Sparkling2009")
metaG_reads <- sweep(metaG_reads, 2, colSums(metaG_reads), "/")

# Add phylum info to keys
# Process metagenome gene key to include a phylum column. Fix any weird formats.

metaG_key$Taxonomy <- gsub("Bacteria;", "", metaG_key$Taxonomy)
metaG_key$Taxonomy <- gsub("Eukaryota;", "", metaG_key$Taxonomy)
metaG_key$Phylum <- sapply(strsplit(as.character(metaG_key$Taxonomy),";"), `[`, 1)

metaG_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("None", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("unclassified", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Unclassified ", "Unclassified", metaG_key$Phylum)
metaG_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", metaG_key$Phylum)
metaG_key$Phylum <- gsub("TM7", "Saccharibacteria", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Ignavibacteriae", "Ignavibacteria", metaG_key$Phylum)
metaG_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", metaG_key$Phylum)
metaG_key$Phylum[which(is.na(metaG_key$Phylum) == T)] <- "Unclassified"
metaG_key$Phylum[grep("Blank", metaG_key$Phylum)] <- "Unclassified"

# Remove unclassified genes to save on RAM
metaG_key <- metaG_key[which(metaG_key$Phylum != "Unclassified"),]

# Split metagenome read count tables by lake and prep for melting
metaG_reads$Genes <- rownames(metaG_reads)
metaG_reads <- metaG_reads[match(metaG_reads$Genes, metaG_key$Gene),]
spark_metaG <- metaG_reads[,c(1,2, 11)]
trout_metaG <- metaG_reads[,c(3,4, 11)]
mendota_metaG <- metaG_reads[,c(5,6, 11)]

# Use melt to switch from wide to long format
spark_metaG <- melt(spark_metaG)
trout_metaG <- melt(trout_metaG)
mendota_metaG <- melt(mendota_metaG)

# Add phylum column to each melted lake table
mendota_metaG$Phylum <- metaG_key$Phylum[match(mendota_metaG$Genes, metaG_key$Gene)]
trout_metaG$Phylum <- metaG_key$Phylum[match(trout_metaG$Genes, metaG_key$Gene)]
spark_metaG$Phylum <- metaG_key$Phylum[match(spark_metaG$Genes, metaG_key$Gene)]

# Aggregate counts by phylum
mendota_metaG_phyla <- aggregate(value ~ Phylum, data = mendota_metaG, mean)
spark_metaG_phyla <- aggregate(value ~ Phylum, data = spark_metaG, mean)
trout_metaG_phyla <- aggregate(value ~ Phylum, data = trout_metaG, mean)

# Write these datasets to file so that I can load them instead of rerunning all of the above when I need to make plot modifications
write.csv(mendota_metaG_phyla, file = paste(path2, "Desktop/intermediate_plotting_files/mendota_metaG_phyla.csv", sep = ""), quote = F, row.names = F)
write.csv(spark_metaG_phyla, file = paste(path2, "Desktop/intermediate_plotting_files/spark_metaG_phyla.csv", sep = ""), quote = F, row.names = F)
write.csv(trout_metaG_phyla, file = paste(path2, "Desktop/intermediate_plotting_files/trout_metaG_phyla.csv", sep = ""), quote = F, row.names = F)

# Remove some files so I have space in RAM to run the next steps
rm(metaG_key)
rm(metaG_reads)
rm(mendota_metaG)
rm(spark_metaG)
rm(trout_metaG)
```

### Main text tables and figures.

**Table 1. Comparison of Sparkling Lake, Lake Mendota, and Trout Bog.** These three lakes were chosen for comparative metatranscriptomics because of their varying trophic statuses, extensive historical data, and previous microbial sampling. Data on surface area, maximum depth, and development on shoreline courtesy of NTL-LTER <lter.limnology.wisc.edu>. Temperature, dissolved oxygen, pH, and conductivity were measured using a HydroLab DS5x Sonde and are averaged over all sampling depths and timepoints for each lake. Chlorophyll and phycocyanin concentrations were measured from the integrated epilimnion samples using a methanol extraction protocol and averaged over all timepoints. Secchi depth was measured at the first timepoint for each lake. Bacterial production was quantified via C14-leucine incorporation and averaged over all timepoints. Due to thunderstorms the night of July 8th, the final 1AM timepoint in Sparkling Lake was collected on July 9th instead. 



|                         | Lake Mendota      | Trout Bog                | Sparkling Lake           |
| ----------------------- |:-----------------:|:------------------------:|:------------------------:|
| Surface area (km^2)     | 39.600              | 0.001                    | 0.637                    |
| Maximum depth (m)       | 25.3              | 7.9                      | 20.o                       |
| Trophic status          | Eutrophic         | Humic                    | Oligotrophic             |
| Location                | Madison, WI USA   | Boulder Junction, WI USA | Boulder Junction, WI USA |
| GPS Coordinates         | 43.1113, -89.4255 | 46.0412, -89.6861        | 46.0091, -89.6695        |
| Shoreline development   | High              | Low                      | Moderate                 |
| Epilimnion sampling depth| 0-7m             | 0-1.5m                   | 0-4m                     |
| Temperature (C)         | `r round(mean(enviro_data$Temperature[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$Temperature[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$Temperature[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
|Dissolved oxygen (mg/L)  | `r round(mean(enviro_data$DO[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$DO[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$DO[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| pH                      | `r round(mean(enviro_data$pH[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$pH[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$pH[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| Conductivity (uS/cm)    | `r round(mean(enviro_data$Conductivity[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data_no_sonde$Conductivity[which(enviro_data_no_sonde$Lake == "TroutBog" & enviro_data_no_sonde$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data_no_sonde$Conductivity[which(enviro_data_no_sonde$Lake == "Sparkling" & enviro_data_no_sonde$Depth <= 4)]), 2)` |
| Total phosphorus (ug/L) | `r round(mean(chemdata$TP[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TP[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TP[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total nitrogen (ug/L)   | `r round(mean(chemdata$TN[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TN[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TN[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total dissolved phosphorus (ug/L) | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TDP[which(chemdata$Lake == "Sparkling")]), 2)` |
| Total dissolved nitrogen (ug/L)   | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Mendota")]), 2)` | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Trout")]), 2)` | `r round(mean(chemdata$TDN[which(chemdata$Lake == "Sparkling")]), 2)` |
| Chlorophyll (ug/L)      | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7 & enviro_data$Timepoint != 40 & enviro_data$Timepoint != 44)]), 2)` | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5 & enviro_data$Timepoint != 20)]), 2)` | `r round(mean(enviro_data$Chlorophyll[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Phycocyanin (ug/L)      | `r round(mean(enviro_data$Phycocyanin[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7 & enviro_data$Timepoint != 40 & enviro_data$Timepoint != 44)]), 2)` | `r round(mean(enviro_data$Phycocyanin[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5 & enviro_data$Timepoint != 20)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|
| Bacterial production (cpm)    | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Mendota" & enviro_data$Depth <= 7)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "TroutBog" & enviro_data$Depth <= 1.5)]), 2)` | `r round(mean(enviro_data$Production[which(enviro_data$Lake == "Sparkling" & enviro_data$Depth <= 4)]), 2)`|    
| Secchi depth (m)        | 4.8               | 1.1                      | 6.2                      |
| Sampling dates (2016)   | July 14-16        | July 8-10                | July 6-9                 |
| Sunrise/sunset time     | 5:32/20:35        | 5:18/20:49               | 5:17/20:50               |

**Table 2.** Gene diffs in Lake Mendota 

```{r, Mendota_table, eval = T, include = F}

##### DESeq differential expression testing

abun_mnorm <- mnorm[order(rowSums(mnorm), decreasing = T), ]
abun_mnorm <- abun_mnorm[1:20000,]

colnames(abun_mnorm) <- gsub(".nonrRNA", "", colnames(abun_mnorm))

input <- as.matrix(abun_mnorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_mnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_mnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)

reskey <- mendota_key[match(rownames(res), mendota_key$Gene), ]

# Write a function to report trends in metabolic categories
pathway.stats <- function(searchterm){
  genes <- reskey$Gene[grep(searchterm, reskey$Product)]
  numgenes <- length(genes)
  gene_results <- res[match(genes, rownames(res)), ]
  gene_counts <- abun_mnorm[match(genes, rownames(abun_mnorm)),]
  gene_results$Day <- rowSums(gene_counts[which(conditions == "day")])
  gene_results$Night <- rowSums(gene_counts[which(conditions == "night")])
  gene_results$Leaning <- gene_results$Day > gene_results$Night
  percent_genes_day <- length(which(gene_results$padj < 0.05 & gene_results$Leaning == TRUE))/length(genes) * 100
  percent_genes_night <- length(which(gene_results$padj < 0.05 & gene_results$Leaning == FALSE))/length(genes) * 100
  percent_reads_day <- sum(gene_results$Day[which(gene_results$padj < 0.05 & gene_results$Leaning == TRUE)])/sum(gene_results$Day) * 100
  percent_reads_night <- sum(gene_results$Night[which(gene_results$padj < 0.05 & gene_results$Leaning == FALSE)])/sum(gene_results$Night) * 100
  day_totals <- colSums(gene_counts[, which(conditions == "day")])
  night_totals <- colSums(gene_counts[which(conditions == "night")])
  sig <- t.test(day_totals, night_totals, alternative = "two.sided")
  
  p.genes <- rownames(gene_results)[which(gene_results$padj < 0.05)]
  if(length(p.genes) > 0){
    p.table <- abun_mnorm[match(p.genes, rownames(abun_mnorm)),]
    p.results <- rain(t(p.table), deltat = 4, period = 24, measure.sequence = table(metadata$Timepoint[match(colnames(p.table), metadata$Sample)]), verbose = T, adjp.method = "Bonferroni")
    percent_cyclic_genes <- length(which(p.results$pVal < 0.05 & p.results$phase == 12))/length(genes) * 100
    gene_results$TotalReads <- apply(gene_results[ ,7:8], 1, sum)
    percent_cyclic_reads <- sum(gene_results$TotalReads[match(p.genes, rownames(gene_results))])/sum(gene_results$TotalReads) * 100
  }else{
    percent_cyclic_genes = 0
    percent_cyclic_reads = 0
  }
  
  report <- c(numgenes, round(percent_genes_day, 2), round(percent_reads_day, 2), round(percent_genes_night, 2), round(percent_reads_night, 2), round(percent_cyclic_genes, 2), round(percent_cyclic_reads, 2), mean(day_totals), mean(night_totals), round(sig$p.value, 2))
  names(report) <- c("%DayGenes", "%DayReadsContributed", "%NightGenes", "%NightReadsContributed", "%CyclicGenes", "%CyclicReadsContributed", "DayReads", "NightReads", "t-test")
  return(report)
}

# Run the function

photo <- pathway.stats("photo|Photo")
rhodo <- pathway.stats("rhodopsin|Rhodopsin")
rubisco <- pathway.stats("rbcL|ribulose-bisphosphate carboxylase")
polyamines <- pathway.stats("putrescine|Putrescine|spermidine|Spermidine")
rTCA <- pathway.stats("citrate lyase|Citrate lyase")
nif <- pathway.stats("nitrogenase|Nitrogenase|NifH|NifD|NifK")
chitinase <- pathway.stats("Chitobiase|chitobiase|chitinase|Chitinase")
ap <- pathway.stats("alkaline phosphatase|Alkaline phosphatase")
cellulase <- pathway.stats("cellulase|cellulose")
ros <- pathway.stats("peroxidase|peroxide|catalase")
nitrite_reduction <- pathway.stats("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase")
protease <- pathway.stats("protease")
carbox <- pathway.stats("carboxylate transport")
ribose <- pathway.stats("ribose transport")
gen.sugars <- pathway.stats("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin")
raff <- pathway.stats("raffinose/stachyose/melibiose transport")
gluc <- pathway.stats("glucose/mannose transport")
rham <- pathway.stats("rhamnose transport")
xyl <- pathway.stats("xylose transport")
fruct <- pathway.stats("fructose transport")
chiti <- pathway.stats("chitobiose transport")
lact <- pathway.stats("lactose/L-arabinose transport")

```


|                         | Number of genes | % Genes more expressed in day      | % Genes more expressed at night   | % Cyclic genes (12 hr phase)  | p-value from t-test of day vs. night read totals | Day/night ratio |
| -----------------------|:-----------------:|:------------------------:|:------------------------:|:------------------------:|:------------------------:|:---:|
| Photosynthesis | `r photo[1]` | `r photo[2]` | `r photo[4]` | `r photo[6]` |`r photo[10]`| `r round(photo[8]/photo[9], 2)` |
| Rhodopsins | `r rhodo[1]` | `r rhodo[2]` | `r rhodo[4]` | `r rhodo[6]` | `r rhodo[10]`|`r round(rhodo[8]/rhodo[9], 2)` |
| RuBisCO | `r rubisco[1]` | `r rubisco[2]` | `r rubisco[4]` | `r rubisco[6]` |`r rubisco[10]` |`r round(rubisco[8]/rubisco[9], 2)` |
| reductive TCA | `r rTCA[1]` | `r rTCA[2]` | `r rTCA[4]` | `r rubisco[6]` | `r rTCA[10]` | `r round(rTCA[8]/rTCA[9], 2)` |
| Polyamines | `r polyamines[1]` | `r polyamines[2]` |  `r polyamines[4]` |  `r polyamines[6]` | `r polyamines[10]` | `r round(polyamines[8]/polyamines[9], 2)` |
| Reactive oxygen species | `r ros[1]` | `r ros[2]` | `r ros[4]` | `r ros[6]` | `r ros[10]` |`r round(ros[8]/ros[9], 2)` |
| Protease | `r protease[1]` | `r protease[2]` | `r protease[4]` | `r protease[6]` | `r protease[10]` | `r round(protease[8]/protease[9], 2)` |
| Ribose transport | `r ribose[1]` | `r ribose[2]` | `r ribose[4]` | `r ribose[6]` |  `r ribose[10]` | `r round(ribose[8]/ribose[9], 2)` |
| General sugar transport | `r gen.sugars[1]` | `r gen.sugars[2]` | `r gen.sugars[4]` | `r gen.sugars[6]` | `r gen.sugars[10]` | `r round(gen.sugars[8]/gen.sugars[9], 2)` |
| Raffinose/stachyose/melibiose transport | `r raff[1]` | `r raff[2]` | `r raff[4]` | `r raff[6]` | `r raff[10]` | `r round(raff[8]/raff[9], 2)` |
| Glucose/mannose transport | `r gluc[1]` | `r gluc[2]` | `r gluc[4]` | `r gluc[6]` | `r gluc[10]` | `r round(gluc[8]/gluc[9], 2)` |
| Rhamnose transport | `r rham[1]` | `r rham[2]` | `r rham[4]` | `r rham[6]` | `r rham[10]` | `r round(rham[8]/rham[9], 2)` |
| Xylose transport | `r xyl[1]` | `r xyl[2]` | `r xyl[4]` | `r xyl[6]` | `r xyl[10]` | `r round(xyl[8]/xyl[9], 2)` |

**Table 3.** Gene diffs in Trout Bog 

```{r, Trout_table, eval = T, include = F}

##### DESeq differential expression testing

abun_tnorm <- tnorm[order(rowSums(tnorm), decreasing = T), ]
abun_tnorm <- abun_tnorm[1:20000,]

colnames(abun_tnorm) <- gsub(".nonrRNA", "", colnames(abun_tnorm))

input <- as.matrix(abun_tnorm)
input <- input/500
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_tnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_tnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)

reskey <- trout_key[match(rownames(res), trout_key$Gene), ]

# Write a function to report trends in metabolic categories
pathway.stats <- function(searchterm){
  genes <- reskey$Gene[grep(searchterm, reskey$Product)]
  numgenes <- length(genes)
  gene_results <- res[match(genes, rownames(res)), ]
  gene_counts <- abun_tnorm[match(genes, rownames(abun_tnorm)),]
  gene_results$Day <- rowSums(gene_counts[which(conditions == "day")])
  gene_results$Night <- rowSums(gene_counts[which(conditions == "night")])
  gene_results$Leaning <- gene_results$Day > gene_results$Night
  percent_genes_day <- length(which(gene_results$padj < 0.05 & gene_results$Leaning == TRUE))/length(genes) * 100
  percent_genes_night <- length(which(gene_results$padj < 0.05 & gene_results$Leaning == FALSE))/length(genes) * 100
  percent_reads_day <- sum(gene_results$Day[which(gene_results$padj < 0.05 & gene_results$Leaning == TRUE)])/sum(gene_results$Day) * 100
  percent_reads_night <- sum(gene_results$Night[which(gene_results$padj < 0.05 & gene_results$Leaning == FALSE)])/sum(gene_results$Night) * 100
  day_totals <- colSums(gene_counts[, which(conditions == "day")])
  night_totals <- colSums(gene_counts[which(conditions == "night")])
  sig <- t.test(day_totals, night_totals, alternative = "two.sided")
  
  p.genes <- rownames(gene_results)[which(gene_results$padj < 0.05)]
  if(length(p.genes) > 0){
    p.table <- abun_tnorm[match(p.genes, rownames(abun_tnorm)),]
    p.results <- rain(t(p.table), deltat = 4, period = 24, measure.sequence = table(metadata$Timepoint[match(colnames(p.table), metadata$Sample)]), verbose = T, adjp.method = "Bonferroni")
    percent_cyclic_genes <- length(which(p.results$pVal < 0.05 & p.results$phase == 12))/length(genes) * 100
    gene_results$TotalReads <- apply(gene_results[ ,7:8], 1, sum)
    percent_cyclic_reads <- sum(gene_results$TotalReads[match(p.genes, rownames(gene_results))])/sum(gene_results$TotalReads) * 100
  }else{
    percent_cyclic_genes = 0
    percent_cyclic_reads = 0
  }
  
  report <- c(numgenes, round(percent_genes_day, 2), round(percent_reads_day, 2), round(percent_genes_night, 2), round(percent_reads_night, 2), round(percent_cyclic_genes, 2), round(percent_cyclic_reads, 2), mean(day_totals), mean(night_totals), round(sig$p.value, 2))
  names(report) <- c("%DayGenes", "%DayReadsContributed", "%NightGenes", "%NightReadsContributed", "%CyclicGenes", "%CyclicReadsContributed", "DayReads", "NightReads", "t-test")
  return(report)
}

# Run the function

photo <- pathway.stats("photo|Photo")
rhodo <- pathway.stats("rhodopsin|Rhodopsin")
rubisco <- pathway.stats("rbcL|ribulose-bisphosphate carboxylase")
polyamines <- pathway.stats("putrescine|Putrescine|spermidine|Spermidine")
rTCA <- pathway.stats("citrate lyase|Citrate lyase")
chitinase <- pathway.stats("Chitobiase|chitobiase|chitinase|Chitinase")
ap <- pathway.stats("alkaline phosphatase|Alkaline phosphatase")
cellulase <- pathway.stats("cellulase|cellulose")
ros <- pathway.stats("peroxidase|peroxide|catalase")
nitrite_reduction <- pathway.stats("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase")
protease <- pathway.stats("protease")
carbox <- pathway.stats("carboxylate transport")
ribose <- pathway.stats("ribose transport")
gen.sugars <- pathway.stats("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin")
raff <- pathway.stats("raffinose/stachyose/melibiose transport")
gluc <- pathway.stats("glucose/mannose transport")
rham <- pathway.stats("rhamnose transport")
xyl <- pathway.stats("xylose transport")
fruct <- pathway.stats("fructose transport")
gh <- pathway.stats("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase")
meth <- pathway.stats("ammonia monooxygenase|methane monoxygenase")
urea <- pathway.stats("urease")
```


|                         | Number of genes | % Genes more expressed in day      | % Genes more expressed at night   | % Cyclic genes (12 hr phase)  | p-value from t-test of day vs. night read totals | Day/night/ratios |
| -----------------------|:-----------------:|:------------------------:|:------------------------:|:------------------------:|:------------------------:|:-----:|
| Photosynthesis | `r photo[1]` | `r photo[2]` | `r photo[4]` | `r photo[6]` | `r photo[10]` | `r round(photo[8]/photo[9], 2)` |
| RuBisCO | `r rubisco[1]` | `r rubisco[2]` | `r rubisco[4]` | `r rubisco[6]` | `r rubisco[10]` | `r round(photo[8]/photo[9], 2)` |
| Glycoside hydrolases | `r gh[1]` | `r gh[2]` | `r gh[4]` | `r gh[6]` | `r gh[10]` | `r round(gh[8]/gh[9], 2)` |
| Polyamines | `r polyamines[1]` | `r polyamines[2]` | `r polyamines[4]` | `r polyamines[6]` | `r polyamines[10]` | `r round(polyamines[8]/polyamines[9], 2)` |
| Reactive oxygen species | `r ros[1]` | `r ros[2]` | `r ros[4]` | `r ros[6]` | `r ros[10]` | `r round(ros[8]/ros[9], 2)` |
| Protease | `r protease[1]` | `r protease[2]` | `r protease[4]` | `r protease[6]` | `r protease[10]` | `r round(protease[8]/protease[9], 2)` |
| Ribose transport | `r ribose[1]` | `r ribose[2]` | `r ribose[4]` | `r ribose[6]` | `r ribose[10]` | `r round(ribose[8]/ribose[9], 2)` |
| General sugar transport | `r gen.sugars[1]` | `r gen.sugars[2]` | `r gen.sugars[4]` | `r gen.sugars[6]` | `r gen.sugars[10]` | `r round(gen.sugars[8]/gen.sugars[9], 2)` |
| Xylose transport | `r xyl[1]` | `r xyl[2]` | `r xyl[4]` | `r xyl[6]` | `r xyl[10]` | `r round(xyl[8]/xyl[9], 2)` |
| Methane/ammonia monooxygenase | `r meth[1]` | `r meth[2]` | `r meth[4]` |  `r meth[6]` | `r meth[10]` | `r round(meth[8]/meth[9], 2)` |


**Table 3.** Gene diffs in Sparkling Lake

```{r, Spark_table, eval = T, include = F}

##### DESeq differential expression testing

abun_snorm <- snorm[order(rowSums(snorm), decreasing = T), ]
abun_snorm <- abun_snorm[1:20000,]

colnames(abun_snorm) <- gsub(".nonrRNA", "", colnames(abun_snorm))

input <- as.matrix(abun_snorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_snorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_snorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)

reskey <- spark_key[match(rownames(res), spark_key$Gene), ]

# Write a function to report trends in metabolic categories
pathway.stats <- function(searchterm){
  genes <- reskey$Gene[grep(searchterm, reskey$Product)]
  numgenes <- length(genes)
  gene_results <- res[match(genes, rownames(res)), ]
  gene_counts <- abun_snorm[match(genes, rownames(abun_snorm)),]
  gene_results$Day <- rowSums(gene_counts[which(conditions == "day")])
  gene_results$Night <- rowSums(gene_counts[which(conditions == "night")])
  gene_results$Leaning <- gene_results$Day > gene_results$Night
  percent_genes_day <- length(which(gene_results$padj < 0.05 & gene_results$Leaning == TRUE))/length(genes) * 100
  percent_genes_night <- length(which(gene_results$padj < 0.05 & gene_results$Leaning == FALSE))/length(genes) * 100
  percent_reads_day <- sum(gene_results$Day[which(gene_results$padj < 0.05 & gene_results$Leaning == TRUE)])/sum(gene_results$Day) * 100
  percent_reads_night <- sum(gene_results$Night[which(gene_results$padj < 0.05 & gene_results$Leaning == FALSE)])/sum(gene_results$Night) * 100
  day_totals <- colSums(gene_counts[, which(conditions == "day")])
  night_totals <- colSums(gene_counts[which(conditions == "night")])
  sig <- t.test(day_totals, night_totals, alternative = "two.sided")
  
  p.genes <- rownames(gene_results)[which(gene_results$padj < 0.05)]
  if(length(p.genes) > 0){
    p.table <- abun_snorm[match(p.genes, rownames(abun_snorm)),]
    p.results <- rain(t(p.table), deltat = 4, period = 24, measure.sequence = table(metadata$Timepoint[match(colnames(p.table), metadata$Sample)]), verbose = T, adjp.method = "Bonferroni")
    percent_cyclic_genes <- length(which(p.results$pVal < 0.05 & p.results$phase == 12))/length(genes) * 100
    gene_results$TotalReads <- apply(gene_results[ ,7:8], 1, sum)
    percent_cyclic_reads <- sum(gene_results$TotalReads[match(p.genes, rownames(gene_results))])/sum(gene_results$TotalReads) * 100
  }else{
    percent_cyclic_genes = 0
    percent_cyclic_reads = 0
  }
  
  report <- c(numgenes, round(percent_genes_day, 2), round(percent_reads_day, 2), round(percent_genes_night, 2), round(percent_reads_night, 2), round(percent_cyclic_genes, 2), round(percent_cyclic_reads, 2), mean(day_totals), mean(night_totals), round(sig$p.value, 2))
  names(report) <- c("%DayGenes", "%DayReadsContributed", "%NightGenes", "%NightReadsContributed", "%CyclicGenes", "%CyclicReadsContributed", "DayReads", "NightReads", "t-test")
  return(report)
}

# Run the function

photo <- pathway.stats("photo|Photo")
rhodo <- pathway.stats("rhodopsin|Rhodopsin")
rubisco <- pathway.stats("rbcL|ribulose-bisphosphate carboxylase")
polyamines <- pathway.stats("putrescine|Putrescine|spermidine|Spermidine")
rTCA <- pathway.stats("citrate lyase|Citrate lyase")
nif <- pathway.stats("nitrogenase|Nitrogenase|NifH|NifD|NifK")
chitinase <- pathway.stats("Chitobiase|chitobiase|chitinase|Chitinase")
ap <- pathway.stats("alkaline phosphatase|Alkaline phosphatase")
cellulase <- pathway.stats("cellulase|cellulose")
ros <- pathway.stats("peroxidase|peroxide|catalase")
nitrite_reduction <- pathway.stats("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase")
protease <- pathway.stats("protease")
carbox <- pathway.stats("carboxylate transport")
ribose <- pathway.stats("ribose transport")
gen.sugars <- pathway.stats("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin")
raff <- pathway.stats("raffinose/stachyose/melibiose transport")
gluc <- pathway.stats("glucose/mannose transport")
rham <- pathway.stats("rhamnose transport")
xyl <- pathway.stats("xylose transport")
fruct <- pathway.stats("fructose transport")
chiti <- pathway.stats("chitobiose transport")
meth <- pathway.stats("ammonia monooxygenase|methane monoxygenase")
urea <- pathway.stats("urease")
```


|                         | Number of genes | % Genes more expressed in day      | % Genes more expressed at night   | % Cyclic genes (12 hr phase)  | p-value from t-test of day vs. night read totals | Day/night ratio |
| -----------------------|:-----------------:|:------------------------:|:------------------------:|:------------------------:|:------------------------:|:---:|
| Photosynthesis | `r photo[1]` | `r photo[2]` | `r photo[4]` | `r photo[6]` | `r photo[10]`| `r round(photo[8]/photo[9], 2)` |
| Rhodopsins | `r rhodo[1]` | `r rhodo[2]` | `r rhodo[4]` | `r rhodo[6]` | `r rhodo[10]`| `r round(rhodo[8]/rhodo[9], 2)` |
| RuBisCO | `r rubisco[1]` | `r rubisco[2]` | `r rubisco[4]` | `r rubisco[6]` | `r rubisco[10]` | `r round(rubisco[8]/rubisco[9], 2)` |
| Polyamines | `r polyamines[1]` | `r polyamines[2]` | `r polyamines[4]` | `r polyamines[6]` | `r polyamines[10]` | `r round(polyamines[8]/polyamines[9], 2)` |
| Alkaline phosphatase | `r ap[1]` | `r ap[2]` | `r ap[4]` | `r ap[6]` | `r ap[10]` | `r round(ap[8]/ap[9], 2)` |
| Reactive oxygen species | `r ros[1]` | `r ros[2]` | `r ros[4]` | `r ros[6]` | `r ros[10]` | `r round(ros[8]/ros[9], 2)` |
| Protease | `r protease[1]` | `r protease[2]` | `r protease[4]` | `r protease[6]` | `r protease[10]` | `r round(protease[8]/protease[9], 2)` |
| Carboxylate transport | `r carbox[1]` | `r carbox[2]` | `r carbox[4]` | `r carbox[6]` | `r carbox[10]` | `r round(carbox[8]/carbox[9], 2)` |
| Ribose transport | `r ribose[1]` | `r ribose[2]` | `r ribose[4]` | `r ribose[6]` | `r ribose[10]` | `r round(ribose[8]/ribose[9], 2)` |
| General sugar transport | `r gen.sugars[1]` | `r gen.sugars[2]` | `r gen.sugars[4]`  | `r gen.sugars[6]` | `r gen.sugars[10]` | `r round(gen.sugars[8]/gen.sugars[9], 2)` |
| Raffinose/stachyose/melibiose transport | `r raff[1]` | `r raff[2]` | `r raff[4]` |  `r raff[6]` | `r raff[10]` | `r round(raff[8]/raff[9], 2)` |
| Glucose/mannose transport | `r gluc[1]` | `r gluc[2]` | `r gluc[4]` | `r gluc[6]` |  `r gluc[10]` | `r round(gluc[8]/gluc[9], 2)` |
| Xylose transport | `r xyl[1]` | `r xyl[2]` | `r xyl[4]` | `r xyl[6]` | `r xyl[10]` | `r round(xyl[8]/xyl[9], 2)` |
| Fructose transport | `r fruct[1]` | `r fruct[2]` | `r fruct[4]` |  `r fruct[6]` | `r fruct[10]` | `r round(fruct[8]/fruct[9], 2)` |



```{r, eval = T, include = F, echo = F}
abun_mnorm <- mnorm[order(rowSums(mnorm), decreasing = T), ]
abun_mnorm <- abun_mnorm[1:20000,]

colnames(abun_mnorm) <- gsub(".nonrRNA", "", colnames(abun_mnorm))

input <- as.matrix(abun_mnorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_mnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_mnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)

reskey <- mendota_key[match(rownames(res), mendota_key$Gene), ]
sig.res.key <- reskey[which(res$padj < 0.05),]

# Make column by certain key words 
sig.res.key$Category <- "None"
sig.res.key$Category[grep("photo|Photo", sig.res.key$Product)] <- "Photosynthesis"
sig.res.key$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", sig.res.key$Product)] <- "Rhodopsin"
# sig.res.key$Category[grep("sugar|Sugar|ribose|Ribose|hexose|maltose|carbohydrate|Carbohydrate|ose transport", sig.res.key$Product)] <- "Sugar degradation"
sig.res.key$Category[grep("rbcL|ribulose-bisphosphate carboxylase", sig.res.key$Product)] <- "RuBisCO"
sig.res.key$Category[grep("putrescine|Putrescine|spermidine|Spermidine", sig.res.key$Product)] <- "Polyamines"
sig.res.key$Category[grep("citrate lyase|Citrate lyase", sig.res.key$Product)] <- "rTCA"
sig.res.key$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", sig.res.key$Product)] <- "Nitrogenase"
sig.res.key$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", sig.res.key$Product)] <- "Chitinase"
sig.res.key$Category[grep("glycoside hydrolase|Glycoside hydrolase|glycosyl hydrolase", sig.res.key$Product)] <- "Glycoside_Hydrolase"
sig.res.key$Category[grep("alkaline phosphatase|Alkaline phosphatase", sig.res.key$Product)] <- "Alkaline_phosphatase"
sig.res.key$Category[grep("cellulase|cellulose", sig.res.key$Product)] <- "Cellulase"
sig.res.key$Category[grep("peroxidase|peroxide|catalase", sig.res.key$Product)] <- "ROS"
sig.res.key$Category[grep("ammonia monooxygenase|methane monoxygenase", sig.res.key$Product)] <- "Methane/Ammonia"
sig.res.key$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", sig.res.key$Product)] <- "Nitrite_reduction"
sig.res.key$Category[grep("urease", sig.res.key$Product)] <- "Urease"
sig.res.key$Category[grep("protease", sig.res.key$Product)] <- "Protease"
sig.res.key$Category[grep("carboxylate transport", sig.res.key$Product)] <- "Carboxylate transport"
sig.res.key$Category[grep("ribose transport", sig.res.key$Product)] <- "Ribose transport"
sig.res.key$Category[grep("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin", sig.res.key$Product)] <- "General sugar transport"
sig.res.key$Category[grep("raffinose/stachyose/melibiose transport", sig.res.key$Product)] <- "Raffinose/stachyose/melibiose transport"
sig.res.key$Category[grep("glucose/mannose transport", sig.res.key$Product)] <- "Glucose/mannose transport"
sig.res.key$Category[grep("rhamnose transport", sig.res.key$Product)] <- "Rhamnose transport"
sig.res.key$Category[grep("xylose transport", sig.res.key$Product)] <- "Xylose transport"
sig.res.key$Category[grep("fructose transport", sig.res.key$Product)] <- "Fructose transport"
sig.res.key$Category[grep("chitobiose transport", sig.res.key$Product)] <- "Chitobiose transport"
sig.res.key$Category[grep("lactose/L-arabinose transport", sig.res.key$Product)] <- "Lactose/arabinose transport"


# Write a function to plot a given category
cycle_plot <- function(type, lake) {
  if(lake == "Mendota") {
    abuntable <- abun_mnorm
  }else if(lake == "Trout"){
    abuntable <- abun_tnorm
  }else if (lake == "Sparkling"){
    abuntable <- abun_snorm
  }
  p.genes <- as.character(sig.res.key$Gene[which(sig.res.key$Category == type)])
  p.table <- abuntable[match(p.genes, rownames(abuntable)),]
  p.results <- rain(t(p.table), deltat = 4, period = 24, measure.sequence = table(metadata$Timepoint[match(colnames(p.table), metadata$Sample)]), verbose = T, adjp.method = "Bonferroni")
 # sig.p.table <- p.table[which(p.results$pVal < 0.05 & p.results$phase == 12), ]
  sig.p.table <- p.table[which(p.results$pVal < 0.05), ]
  sig.p.table <- as.data.frame(t(apply(sig.p.table, 1, zscore)))
  sig.p.table$Genes <- rownames(sig.p.table)
  melt.sig.p.table <- melt(sig.p.table)
  melt.sig.p.table$Timepoint <- metadata$Timepoint[match(melt.sig.p.table$variable, metadata$Sample)]
  ave.sig.p.table <- aggregate(value ~ Genes + Timepoint, melt.sig.p.table, mean)
  ave.sig.p.table$phase <- p.results$phase[match(ave.sig.p.table$Genes, rownames(p.results))]
  ave.sig.p.table$phase <- factor(ave.sig.p.table$phase, levels = c("4", "8", "12", "16", "20", "24"))
  day.night.boxes <- data.frame(x1 = c(0, 0.53, 15.35, 24.53, 39.35), x2 = c(0.53, 15.53, 24.53, 39.53, 44), y1 = rep(min(ave.sig.p.table$value), 5), y2 = rep(max(ave.sig.p.table$value), 5), labels = c("", "Sunrise", "Sunset", "Sunrise", "Sunset"))
  #p <- ggplot()  + geom_rect(data = day.night.boxes, aes(xmin = x1, xmax = x2, ymin = y1, ymax = y2), fill = c("lightsteelblue", "lemonchiffon", "lightsteelblue", "lemonchiffon", "lightsteelblue"), alpha = 0.5) + geom_text(data = day.night.boxes, inherit.aes = FALSE, aes(x = x1 + 2, y = y2 - 0.25, label = labels)) + geom_line(data = ave.sig.p.table, inherit.aes = F, aes(x = Timepoint, y = value, group = Genes), color = "grey") + stat_summary(data = ave.sig.p.table, inherit.aes = F, aes(x = Timepoint, y = value), fun.y=mean, geom="line", colour="red") + scale_x_continuous(breaks = seq(0,44, by = 4)) + labs(x = "Hours into Time Series", y = "Z-score normalized reads", title = paste(lake, type)) + background_grid(major = "xy", minor = "none")
  
colors <- c("#4575b4", "#d73027", "#fc8d59", "#fee090", "#e0f3f8", "#91bfdb")
phases <- c("4", "8", "12", "16", "20", "24")
  p <- ggplot()  + geom_rect(data = day.night.boxes, aes(xmin = x1, xmax = x2, ymin = y1, ymax = y2), fill = c("lightsteelblue", "lemonchiffon", "lightsteelblue", "lemonchiffon", "lightsteelblue"), alpha = 0.5) + geom_text(data = day.night.boxes, inherit.aes = FALSE, aes(x = x1 + 2, y = y2 - 0.25, label = labels)) + geom_line(data = ave.sig.p.table, inherit.aes = F, aes(x = Timepoint, y = value, group = Genes, color = phase)) + scale_x_continuous(breaks = seq(0,44, by = 4)) + labs(x = "Hours into Time Series", y = "Z-score normalized reads", title = paste(lake, type)) + background_grid(major = "xy", minor = "none") + scale_color_manual(values = colors[which(phases %in% unique(ave.sig.p.table$phase))])
  return(p)
}

cycle1 <- cycle_plot("Photosynthesis", "Mendota")
cycle2 <- cycle_plot("ROS", "Mendota")
cycle3 <- cycle_plot("Rhodopsin", "Mendota")
```

```{r, eval = T, echo = F}
# calculate here but save plots for later
cycle1
cycle2
cycle3
```
**Figure 1. Cyclic trends in Lake Mendota.** Cyclic trends with a 12 hour phase were detected in Lake Mendota for several functional categories. 


```{r, eval = T, warning = F, message = F, echo = F}
##### DESeq differential expression testing

abun_mnorm <- mnorm[order(rowSums(mnorm), decreasing = T), ]
abun_mnorm <- abun_mnorm[1:20000,]

colnames(abun_mnorm) <- gsub(".nonrRNA", "", colnames(abun_mnorm))

input <- as.matrix(abun_mnorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_mnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_mnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)

res <- res[which(res$padj < 0.05), ]
reskey <- mendota_key[match(rownames(res), mendota_key$Gene), ]
reskey$log2FoldChange <- res$log2FoldChange[match(reskey$Gene, rownames(res))]

# Make column by certain key words 
reskey$Category <- "None"
reskey$Category[grep("photo|Photo", reskey$Product)] <- "Photosynthesis"
reskey$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", reskey$Product)] <- "Rhodopsin"
#reskey$Category[grep("rbcL|ribulose-bisphosphate carboxylase", reskey$Product)] <- "RuBisCO"
#reskey$Category[grep("citrate lyase|Citrate lyase", reskey$Product)] <- "rTCA"
#reskey$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", reskey$Product)] <- "Nitrogenase"
reskey$Category[grep("peroxidase|peroxide|catalase", reskey$Product)] <- "ROS"
#reskey$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", reskey$Product)] <- "Nitrite_reduction"
#reskey$Category[grep("urease", reskey$Product)] <- "Urease"
reskey$Category[grep("protease", reskey$Product)] <- "Protease"
reskey$Category[grep("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin", reskey$Product)] <- "General sugar transport"
reskey$Category[grep("raffinose/stachyose/melibiose transport", reskey$Product)] <- "Raffinose/stachyose/melibiose transport"
#reskey$Category[grep("chitobiose transport", reskey$Product)] <- "Chitobiose transport"
#reskey$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", reskey$Product)] <- "Chitinase"
reskey$Category[grep("ribose transport", reskey$Product)] <- "Ribose transport"

# Add day/night counts
gene_counts <- abun_mnorm[match(reskey$Gene, rownames(abun_mnorm)),]
reskey1 <- reskey
reskey2 <- reskey
reskey1$Counts <- rowSums(gene_counts[which(conditions == "day")])
reskey2$Counts <- rowSums(gene_counts[which(conditions == "night")])
reskey <- rbind(reskey1, reskey2)
reskey$Time <- c(rep("Day", dim(reskey1)[1]), rep("Night", dim(reskey2)[1]))
reskey$Counts <- as.numeric(reskey$Counts)

# Visually inspect phyla to keep
# ggplot(data = reskey, aes(x = Phylum, y = Counts, fill = Time)) + geom_bar(stat = "identity")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Mendota") + facet_wrap(~ Category, nrow = 3, scales="free")

phyla2keep <- c("Actinobacteria", "Bacteroidetes", "Unclassified", "Cyanobacteria", "Betaproteobacteria", "Gammaproteobacteria", "Verrucomicrobia", "Firmicutes", "Deltaproteobacteria", "Planctomycetes")
reskey <- reskey[which(reskey$Category != "None"), ] 
`%not in%` <- function (x, table) is.na(match(x, table, nomatch=NA_integer_))
reskey$Phylum[which(is.na(reskey$Phylum) == T)] <- "Unclassified"
reskey$Phylum <- as.character(reskey$Phylum)
reskey$Phylum[which(reskey$Phylum %not in% phyla2keep)] <- "Other"


colorkey <- data.frame(phylum = c("Actinobacteria", "Alphaproteobacteria", "Armatimonadetes", "Bacteroidetes", "Betaproteobacteria", "Cyanobacteria", "Deltaproteobacteria", "Firmicutes", "Gammaproteobacteria", "Other", "Planctomycetes", "Unclassified", "Verrucomicrobia"), color = c("chocolate1",  "hotpink2", "turquoise3", "goldenrod1", "dodgerblue", "green3", "skyblue", "orchid3", "palegreen", "honeydew4", "lightcoral", "honeydew3", "lightslateblue"))

ggplot(data = reskey, aes(x = Time, y = Counts, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Mendota") + facet_wrap(~ Category, nrow = 3) + scale_fill_manual(values = as.character(colorkey$color[which(colorkey$phylum %in% unique(reskey$Phylum))]))
```



```{r, eval = T, warning = F, message = F, echo = F}

##### DESeq differential expression testing

abun_tnorm <- tnorm[order(rowSums(tnorm), decreasing = T), ]
abun_tnorm <- abun_tnorm[1:20000,]

colnames(abun_tnorm) <- gsub(".nonrRNA", "", colnames(abun_tnorm))

input <- as.matrix(abun_tnorm)
input <- input/500
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_tnorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_tnorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)
res <- res[which(res$padj < 0.05), ]
reskey <- trout_key[match(rownames(res), trout_key$Gene), ]
reskey$log2FoldChange <- res$log2FoldChange[match(reskey$Gene, rownames(res))]

# Make column by certain key words 
reskey$Category <- "None"
reskey$Category[grep("photo|Photo", reskey$Product)] <- "Photosynthesis"
#reskey$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", reskey$Product)] <- "Rhodopsin"
reskey$Category[grep("rbcL|ribulose-bisphosphate carboxylase", reskey$Product)] <- "RuBisCO"
#reskey$Category[grep("citrate lyase|Citrate lyase", reskey$Product)] <- "rTCA"
#reskey$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", reskey$Product)] <- "Nitrogenase"
reskey$Category[grep("peroxidase|peroxide|catalase", reskey$Product)] <- "ROS"
#reskey$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", reskey$Product)] <- "Nitrite_reduction"
#reskey$Category[grep("urease", reskey$Product)] <- "Urease"
#reskey$Category[grep("protease", reskey$Product)] <- "Protease"
reskey$Category[grep("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin", reskey$Product)] <- "General sugar transport"
#reskey$Category[grep("raffinose/stachyose/melibiose transport", reskey$Product)] <- "Raffinose/stachyose/melibiose transport"
#reskey$Category[grep("chitobiose transport", reskey$Product)] <- "Chitobiose transport"
#reskey$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", reskey$Product)] <- "Chitinase"
reskey$Category[grep("ribose transport", reskey$Product)] <- "Ribose transport"
#reskey$Category[grep("glucose/mannose transport", reskey$Product)] <- "Glucose/mannose transport"
#reskey$Category[grep("rhamnose transport", reskey$Product)] <- "Rhamnose transport"
reskey$Category[grep("xylose transport", reskey$Product)] <- "Xylose transport"
#reskey$Category[grep("fructose transport", reskey$Product)] <- "Fructose transport"

# Add day/night counts
gene_counts <- abun_mnorm[match(reskey$Gene, rownames(abun_tnorm)),]
reskey1 <- reskey
reskey2 <- reskey
reskey1$Counts <- rowSums(gene_counts[which(conditions == "day")])
reskey2$Counts <- rowSums(gene_counts[which(conditions == "night")])
reskey <- rbind(reskey1, reskey2)
reskey$Time <- c(rep("Day", dim(reskey1)[1]), rep("Night", dim(reskey2)[1]))
reskey$Counts <- as.numeric(reskey$Counts)

# Visually inspect phyla to keep
# ggplot(data = reskey, aes(x = Phylum, y = Counts, fill = Time)) + geom_bar(stat = "identity")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Mendota") + facet_wrap(~ Category, nrow = 3, scales="free")

phyla2keep <- c("Actinobacteria", "Alphaproteobacteria", "Armatimonadetes", "Bacteroidetes", "Unclassified", "Cyanobacteria", "Betaproteobacteria", "Gammaproteobacteria", "Verrucomicrobia", "Deltaproteobacteria", "Eukaryota", "Chlorophyta")
reskey <- reskey[which(reskey$Category != "None"), ] 
`%not in%` <- function (x, table) is.na(match(x, table, nomatch=NA_integer_))
reskey$Phylum[which(is.na(reskey$Phylum) == T)] <- "Unclassified"
reskey$Phylum <- as.character(reskey$Phylum)
reskey$Phylum[which(reskey$Phylum %not in% phyla2keep)] <- "Other"


colorkey <- data.frame(phylum = c("Actinobacteria", "Alphaproteobacteria", "Armatimonadetes", "Bacteroidetes", "Betaproteobacteria", "Chlorophyta", "Cyanobacteria", "Deltaproteobacteria", "Eukaryota", "Firmicutes", "Gammaproteobacteria", "Other", "Planctomycetes", "Unclassified", "Verrucomicrobia"), color = c("chocolate1",  "hotpink2", "turquoise3", "goldenrod1", "dodgerblue", "palegreen4", "green3", "skyblue", "darkolivegreen3", "orchid3", "palegreen", "honeydew4", "lightcoral", "honeydew3", "lightslateblue"))

ggplot(data = reskey, aes(x = Time, y = Counts, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Trout Bog") + facet_wrap(~ Category, nrow = 3) + scale_fill_manual(values = as.character(colorkey$color[which(colorkey$phylum %in% unique(reskey$Phylum))]))
```

```{r, eval = T, warning = F, message = F, echo = F}
##### DESeq differential expression testing

abun_snorm <- snorm[order(rowSums(snorm), decreasing = T), ]
abun_snorm <- abun_snorm[1:20000,]

colnames(abun_snorm) <- gsub(".nonrRNA", "", colnames(abun_snorm))

input <- as.matrix(abun_snorm)
input <- input/100
input <- round(input, digits = 0)

conditions <- metadata$Time[match(colnames(abun_snorm), metadata$Sample)]
conditions[which(conditions == 9 | conditions == 13 | conditions == 17)] <- "day"
conditions[which(conditions == 5 | conditions == 21 | conditions == 1)] <- "night"
coldata <- data.frame(samples = colnames(abun_snorm), conditions)

cds <- DESeqDataSetFromMatrix(countData = input,
                              colData = coldata,
                              design = ~ conditions)

cds <- estimateSizeFactors(cds) 
cds <- estimateDispersions(cds) 


dds <- DESeq(cds)
res <- results(dds)

res <- res[which(res$padj < 0.05), ]
reskey <- spark_key[match(rownames(res), spark_key$Gene), ]
reskey$log2FoldChange <- res$log2FoldChange[match(reskey$Gene, rownames(res))]

# Make column by certain key words 
reskey$Category <- "None"
reskey$Category[grep("photo|Photo", reskey$Product)] <- "Photosynthesis"
#reskey$Category[grep("rhodopsin|Rhodopsin|phytoene|lycopene|carotene|Phytoene|Lycopene|Carotene", reskey$Product)] <- "Rhodopsin"
#reskey$Category[grep("rbcL|ribulose-bisphosphate carboxylase", reskey$Product)] <- "RuBisCO"
#reskey$Category[grep("citrate lyase|Citrate lyase", reskey$Product)] <- "rTCA"
#reskey$Category[grep("nitrogenase|Nitrogenase|NifH|NifD|NifK", reskey$Product)] <- "Nitrogenase"
#reskey$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", reskey$Product)] <- "Chitinase"
reskey$Category[grep("peroxidase|peroxide|catalase", reskey$Product)] <- "ROS"
#reskey$Category[grep("nitrite reductase|Nitrite reductase|nitrite oxidoreductase|Nitrite oxidoreductase", reskey$Product)] <- "Nitrite_reduction"
#reskey$Category[grep("urease", reskey$Product)] <- "Urease"
#reskey$Category[grep("protease", reskey$Product)] <- "Protease"
#reskey$Category[grep("sugar transport|carbohydrate ABC transport|Carbohydrate-selective porin", reskey$Product)] <- "General sugar transport"
reskey$Category[grep("raffinose/stachyose/melibiose transport", reskey$Product)] <- "Raffinose/stachyose/melibiose transport"
#reskey$Category[grep("chitobiose transport", reskey$Product)] <- "Chitobiose transport"
#reskey$Category[grep("Chitobiase|chitobiase|chitinase|Chitinase", reskey$Product)] <- "Chitinase"
#reskey$Category[grep("ribose transport", reskey$Product)] <- "Ribose transport"

# Add day/night counts
gene_counts <- abun_snorm[match(reskey$Gene, rownames(abun_snorm)),]
reskey1 <- reskey
reskey2 <- reskey
reskey1$Counts <- rowSums(gene_counts[which(conditions == "day")])
reskey2$Counts <- rowSums(gene_counts[which(conditions == "night")])
reskey <- rbind(reskey1, reskey2)
reskey$Time <- c(rep("Day", dim(reskey1)[1]), rep("Night", dim(reskey2)[1]))
reskey$Counts <- as.numeric(reskey$Counts)

# Visually inspect phyla to keep
# ggplot(data = reskey, aes(x = Phylum, y = Counts, fill = Time)) + geom_bar(stat = "identity")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Sparkling") + facet_wrap(~ Category, nrow = 3, scales="free")

phyla2keep <- c("Actinobacteria", "Bacteroidetes", "Unclassified", "Cyanobacteria", "Betaproteobacteria", "Gammaproteobacteria", "Verrucomicrobia", "Firmicutes", "Deltaproteobacteria", "Planctomycetes", "Eukaryota")
reskey <- reskey[which(reskey$Category != "None"), ] 
`%not in%` <- function (x, table) is.na(match(x, table, nomatch=NA_integer_))
reskey$Phylum[which(is.na(reskey$Phylum) == T)] <- "Unclassified"
reskey$Phylum <- as.character(reskey$Phylum)
reskey$Phylum[which(reskey$Phylum %not in% phyla2keep)] <- "Other"


colorkey <- data.frame(phylum = c("Actinobacteria", "Alphaproteobacteria", "Armatimonadetes", "Bacteroidetes", "Betaproteobacteria", "Chlorophyta", "Cyanobacteria", "Deltaproteobacteria", "Eukaryota", "Firmicutes", "Gammaproteobacteria", "Other", "Planctomycetes", "Unclassified", "Verrucomicrobia"), color = c("chocolate1",  "hotpink2", "turquoise3", "goldenrod1", "dodgerblue", "palegreen4", "green3", "skyblue", "darkolivegreen3", "orchid3", "palegreen", "honeydew4", "lightcoral", "honeydew3", "lightslateblue"))

ggplot(data = reskey, aes(x = Time, y = Counts, fill = Phylum)) + geom_bar(stat = "identity", position = "fill")  + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_y_continuous(expand = c(0, 0)) + labs(x = NULL, y = "Proportion of reads assigned", title = "Sparkling") + facet_wrap(~ Category, nrow = 3) + scale_fill_manual(values = as.character(colorkey$color[which(colorkey$phylum %in% unique(reskey$Phylum))]))
```


```{r, eval = T, echo = F}
totals <- rowSums(snorm)
top10 <- rownames(snorm)[order(totals, decreasing = T)]
top10 <- top10[1:11]
top10 <- spark_key[match(top10, spark_key$Gene),]
top10$totals <- match(top10$Gene, names(totals))
top10 <- top10[which(top10$Product != "internal standard"), ]
top10$Phylum[which(is.na(top10$Phylum) == T)] <- "Unclassified"
top10$Gene <- factor(top10$Gene, levels = top10$Gene[order(top10$totals, decreasing = T)])
p1 <- ggplot(top10, aes(x = Gene, y = totals, color = Phylum)) + geom_point(size = 5) + theme(axis.text.x = element_blank()) + geom_text_repel(aes(label = top10$Product[order(top10$totals, decreasing = T)]), force = 10, size = 4, color = "black", alpha = 0.5) + labs(x = "Gene Rank Abundance", y = "Summed Normalized Read Counts", title = "Sparkling Lake Top 10 Expressed Genes")

totals.2 <- rowSums(mnorm)
top10.2 <- rownames(mnorm)[order(totals.2, decreasing = T)]
top10.2 <- top10.2[1:11]
top10.2 <- mendota_key[match(top10.2, mendota_key$Gene),]
top10.2$totals.2 <- match(top10.2$Gene, names(totals.2))
top10.2 <- top10.2[which(top10.2$Product != "internal standard"), ]
top10.2$Phylum[which(is.na(top10.2$Phylum) == T)] <- "Unclassified"
top10.2$Gene <- factor(top10.2$Gene, levels = top10.2$Gene[order(top10.2$totals.2, decreasing = T)])
p2 <- ggplot(top10.2, aes(x = Gene, y = totals.2, color = Phylum)) + geom_point(size = 5) + theme(axis.text.x = element_blank()) + geom_text_repel(aes(label = top10.2$Product[order(top10.2$totals.2, decreasing = T)]), force = 10, size = 4, color = "black", alpha = 0.5) + labs(x = "Gene Rank Abundance", y = "Summed Normalized Read Counts", title = "Lake Mendota Top 10 Expressed Genes")

totals.3 <- rowSums(tnorm)
top10.3 <- rownames(tnorm)[order(totals.3, decreasing = T)]
top10.3 <- top10.3[1:11]
top10.3 <- trout_key[match(top10.3, trout_key$Gene),]
top10.3$totals.3 <- match(top10.3$Gene, names(totals.3))
top10.3 <- top10.3[which(top10.3$Product != "internal standard"), ]
top10.3$Phylum[which(is.na(top10.3$Phylum) == T)] <- "Unclassified"
top10.3$Gene <- factor(top10.3$Gene, levels = top10.3$Gene[order(top10.3$totals.3, decreasing = T)])
p3 <- ggplot(top10.3, aes(x = Gene, y = totals.3, color = Phylum)) + geom_point(size = 5) + theme(axis.text.x = element_blank()) + geom_text_repel(aes(label = top10.3$Product[order(top10.3$totals.3, decreasing = T)]), force = 10, size = 4, color = "black", alpha = 0.5) + labs(x = "Gene Rank Abundance", y = "Summed Normalized Read Counts", title = "Trout Bog Top 10 Expressed Genes")

#fig1 <- plot_grid(p1, p2, p3, nrow = 3)
p1
p2
p3
```

**Supp Figure . Top 10 most expressed annotated heterotrophic genes in each study site.** Coding regions from reference genomes and metagenome assemblies were clustered at 97% sequence similarity, and the longest coding region was chosen as the representative sequence. Metatranscriptomic reads were mapped to these representative sequences. The top 10 most expressed genes from each lake, filtered to exclude photosynthetic genes, phototrophic organisms, hypothetical genes, and unclassified genes, are presented here. Annotations and classifications are derived from the sequence to which each read mapped.    


```{r, eval = T, echo = F, fig.height = 18, fig.width = 18, warning = F, message = F}

mendota_metaG_phyla <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/mendota_metaG_phyla.csv", sep = ""), header = T, colClasses = c("character", "numeric"))
spark_metaG_phyla <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/spark_metaG_phyla.csv", sep = ""), header = T, colClasses = c("character", "numeric"))
trout_metaG_phyla <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/trout_metaG_phyla.csv", sep = ""), header = T, colClasses = c("character", "numeric"))
# 
# 
# # Include a key for colorcoding phyla by broader classification
# 
type_key <- data.frame(phylum = c("Actinobacteria", "Bacteroidetes", "Chloroflexi", "Crenarchaeota", "Cryptophyta", "Cyanobacteria", "Gemmatimonadetes", "Heterokonta", "Ignavibacteria", "Planctomycetes", "Proteobacteria", "Verrucomicrobia", "Viruses", "Armatimonadetes", "Firmicutes", "Ciliophora", "Acidobacteria", "Tenericutes", "Arthropoda", "Chlorobi", "Candidatus Saccharibacteria", "Chlorophyta", "Deinococcus-Thermus", "Elusimicrobia", "Haptophyta", "Spirochaetes", "Phaeophyceae", "Streptophyta"), type = c("Bacteria", "Bacteria", "Bacteria", "Archaea", "Algae", "Bacteria", "Bacteria", "Algae", "Bacteria", "Bacteria", "Bacteria", "Bacteria", "Viruses", "Bacteria", "Bacteria", "Protist", "Bacteria", "Bacteria", "Animal", "Bacteria", "Bacteria", "Algae", "Bacteria", "Bacteria", "Algae", "Bacteria", "Algae", "Algae"))


#Process Mendota expression data


# # Remove unclassified genes and the internal standard
# mendota_key <- mendota_key[grep("Unclassified|internal standard", mendota_key$Phylum, invert = T), ]
# 
# # Prep read count table and switch to long format
# mnorm2 <- mnorm
# mnorm2$Genes <- rownames(mnorm2)
# mnorm2 <- mnorm2[which(mnorm2$Genes %in% mendota_key$Gene), ]
# colnames(mnorm2) <- gsub(".nonrRNA", "", colnames(mnorm2))
# mnorm2 <- melt(mnorm2)
# # Add columns of timepoint and phylum
# mnorm2$Timepoint <- metadata$Timepoint[match(mnorm2$variable, metadata$Sample)]
# mnorm2$Taxonomy <- mendota_key$Phylum[match(mnorm2$Genes, mendota_key$Gene)]
# 
# # Aggregate by phylum
# mendota_phyla <- aggregate(value ~ Taxonomy, data = mnorm2, mean)
# 
# # Save an intermediate file of Mendota data
# write.csv(mendota_phyla, file = paste(path2, "Desktop/intermediate_plotting_files/mendota_metaT_phyla.csv", sep = ""), quote = F, row.names = F)
# Read back in
mendota_phyla <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/mendota_metaT_phyla.csv", sep = ""), header = T, colClasses = c("character", "numeric"))

# Keep the top half of both expressed and abundant phyla for plotting
# keep_phyla <- c(mendota_metaG_phyla$Phylum[which(mendota_metaG_phyla$value > quantile(mendota_metaG_phyla$value)[3])], mendota_phyla$Taxonomy[which(mendota_phyla$value > quantile(mendota_phyla$value)[3])])
# mendota_phyla <- mendota_phyla[which(mendota_phyla$Taxonomy %in% keep_phyla),]
mendota_phyla$metaG <- mendota_metaG_phyla$value[match(mendota_phyla$Taxonomy, mendota_metaG_phyla$Phylum)]
mendota_phyla$type <- type_key$type[match(mendota_phyla$Taxonomy, type_key$phylum)]
mendota_phyla <- mendota_phyla[which(is.na(mendota_phyla$metaG) == F),]



# Color key: c("limegreen", "lavenderblush4", "royalblue", "goldenrod")) == algae, archaea, bacteria, viruses
p1 <- ggplot(mendota_phyla[which(mendota_phyla$Taxonomy != "Chloroflexi"), ], aes(x = metaG, y = value)) + geom_point(size = 2.5) + geom_abline(intercept = 0, slope = max(mendota_phyla$value)/max(mendota_phyla$metaG), linetype = 3) + geom_text_repel(aes(label = Taxonomy), force = 10, size = 3, color = "black") + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean transcripts/L assigned", title = "Lake Mendota") + theme(legend.position = "none") + scale_x_continuous(limits = c(0, max(mendota_phyla$metaG)))


#Process Sparkling expression data

# Remove unclassified genes and the internal standard
 spark_key <- spark_key[grep("Unclassified|internal standard", spark_key$Phylum, invert = T), ]
# 
# # Prep read count table and switch to long format
# snorm2 <- snorm
# snorm2$Genes <- rownames(snorm2)
# snorm2 <- snorm2[which(snorm2$Genes %in% spark_key$Gene), ]
# colnames(snorm2) <- gsub(".nonrRNA", "", colnames(snorm2))
# snorm2 <- melt(snorm2)
# # Add columns of timepoint and phylum
# snorm2$Timepoint <- metadata$Timepoint[match(snorm2$variable, metadata$Sample)]
# snorm2$Taxonomy <- spark_key$Phylum[match(snorm2$Genes, spark_key$Gene)]

# Aggregate by phylum
# spark_phyla <- aggregate(value ~ Taxonomy, data = snorm2, mean)

# Save an intermediate file of Mendota data
# write.csv(spark_phyla, file = paste(path2, "Desktop/intermediate_plotting_files/spark_metaT_phyla.csv", sep = ""), quote = F, row.names = F)
# Read back in
spark_phyla <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/spark_metaT_phyla.csv", sep = ""), header = T, colClasses = c("character", "numeric"))


# Keep the top half of both expressed and abundant phyla for plotting
# keep_phyla <- c(spark_metaG_phyla$Phylum[which(spark_metaG_phyla$value > quantile(spark_metaG_phyla$value)[3])], spark_phyla$Taxonomy[which(spark_phyla$value > quantile(spark_phyla$value)[3])])
# spark_phyla <- spark_phyla[which(spark_phyla$Taxonomy %in% keep_phyla),]
spark_phyla$metaG <- spark_metaG_phyla$value[match(spark_phyla$Taxonomy, spark_metaG_phyla$Phylum)]
spark_phyla$type <- type_key$type[match(spark_phyla$Taxonomy, type_key$phylum)]
spark_phyla <- spark_phyla[which(is.na(spark_phyla$metaG) == F),]

p2 <- ggplot(spark_phyla, aes(x = metaG, y = value)) + geom_point(size = 2.5) + geom_abline(intercept = 0, slope = max(spark_phyla$value)/max(spark_phyla$metaG), linetype = 3) + geom_text_repel(aes(label = Taxonomy), force = 10, size = 3, color = "black") + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean transcripts/L assigned", title = "Sparkling Lake") + theme(legend.position = "none") + scale_x_continuous(limits = c(0, max(spark_phyla$metaG)))
#Process Trout expression data

# # # Remove unclassified genes and the internal standard
trout_key <- trout_key[grep("Unclassified|internal standard", trout_key$Phylum, invert = T), ]

# Prep read count table and switch to long format
# tnorm2 <- tnorm
# tnorm2$Genes <- rownames(tnorm2)
# tnorm2 <- tnorm2[which(tnorm2$Genes %in% trout_key$Gene), ]
# colnames(tnorm2) <- gsub(".nonrRNA", "", colnames(tnorm2))
# tnorm2 <- melt(tnorm2)
# # Add columns of timepoint and phylum
# tnorm2$Timepoint <- metadata$Timepoint[match(tnorm2$variable, metadata$Sample)]
# tnorm2$Taxonomy <- trout_key$Phylum[match(tnorm2$Genes, trout_key$Gene)]
# 
# # Aggregate by phylum
# trout_phyla <- aggregate(value ~ Taxonomy, data = tnorm2, mean)
# 
# # Save an intermediate file of Mendota data
# write.csv(trout_phyla, file = paste(path2, "Desktop/intermediate_plotting_files/trout_metaT_phyla.csv", sep = ""), quote = F, row.names = F)
# Read back in
trout_phyla <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/trout_metaT_phyla.csv", sep = ""), header = T, colClasses = c("character", "numeric"))

# Keep the top half of both expressed and abundant phyla for plotting
# keep_phyla <- c(trout_metaG_phyla$Phylum[which(trout_metaG_phyla$value > quantile(trout_metaG_phyla$value)[3])], trout_phyla$Taxonomy[which(trout_phyla$value > quantile(trout_phyla$value)[3])])
# trout_phyla <- trout_phyla[which(trout_phyla$Taxonomy %in% keep_phyla),]
trout_phyla$metaG <- trout_metaG_phyla$value[match(trout_phyla$Taxonomy, trout_metaG_phyla$Phylum)]
trout_phyla$type <- type_key$type[match(trout_phyla$Taxonomy, type_key$phylum)]
trout_phyla <- trout_phyla[which(is.na(trout_phyla$metaG) == F),]

p3 <- ggplot(trout_phyla, aes(x = metaG, y = value)) + geom_point(size = 2.5) + geom_abline(intercept = 0, slope = max(trout_phyla$value)/max(trout_phyla$metaG), linetype = 3) + geom_text_repel(aes(label = Taxonomy), force = 10, size = 3, color = "black") + labs(x = "Mean proportion of metagenomic reads assigned", y = "Mean transcripts/L assigned", title = "Trout Bog") + theme(legend.position = "none") + scale_x_continuous(limits = c(0, max(trout_phyla$metaG)))

#plot_grid(p1, p2, p3, nrow = 3, labels = c("A", "B", "C"))
p1
p2
p3

# Remove datasets to save space
#rm(snorm2)
#rm(tnorm2)
```

**Supp Figure. Most highly expressed or abundant phyla by lake.** To determine which phyla were most abundant or most expressed during our time series, we analyzed metagenomic and metatranscriptomic read counts. The expression of clustered, nonredundant genes was aggregated by phylum and compared to the coverage of those phyla in metagenomes. Genes that could not be classified in a phylum were not included in this analysis. No positive relationship was observed between expression and abundance. Color indicates the type of phylum represented, with grey indicating Archaea, blue for Bacteria, green for Algae, red for multicellular animals, black for protists, and yellow for viruses. The identity of the most expressed and most abundant phyla varied by lake. One phylum, Chloroflexi, was removed from the plot of Lake Mendota due to orders of magnitude higher expression and abundance. This phylum is likely an outlier.

```{r, eval = T, echo = F, fig.height = 5, fig.width = 15, message = F, warning = F}
# Write a function to calculate CoV within replicates and average by gene
within_reps <- function(readcounts){
  readcounts2 <- readcounts[which(rowSums(readcounts) > 1000), ]
  colnames(readcounts2) <- gsub(".nonrRNA", "", colnames(readcounts2))
  timepoints <- metadata$Timepoint[match(colnames(readcounts2), metadata$Sample)]
  timelist <- unique(timepoints)
  time_table <- readcounts2[,which(timepoints == timelist[1])]
  results <- data.frame(timepoint1 = apply(time_table, 1, cv))

  for(i in 2:length(timelist)){
    if(length(which(timepoints == timelist[i])) > 1){
      time_table <- readcounts2[,which(timepoints == timelist[i])]
      next_timepoint <- data.frame(apply(time_table, 1, cv))
      colnames(next_timepoint) <- paste("timepoint", i, sep = "")
      results <- cbind(results, next_timepoint)
    }
  }
  return(results)
}

mendota_within <- within_reps(mnorm)
mendota_within <- apply(mendota_within, 1, mean, na.rm = T)
mendota_across <- apply(mnorm, 1, cv, na.rm = T)
cv_stat <- c(rep("within replicates", length(mendota_within)), rep("across samples", length(mendota_across)))
mendota_cv_stats <- data.frame(type = cv_stat, stat = c(mendota_within, mendota_across))

# Save data to avoid rerunning the above code
write.csv(mendota_cv_stats, file = paste(path2, "Desktop/intermediate_plotting_files/mendota_cv_stats.csv", sep = ""), quote = F, row.names = F)

# Read data back in
mendota_cv_stats <- read.csv(file = paste(path2, "Desktop/intermediate_plotting_files/mendota_cv_stats.csv", sep = ""), header = T)

p1 <- ggplot(mendota_cv_stats, aes(stat, fill = type)) + geom_density(alpha = 0.5) + scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + labs(x = "Coefficent of Variance (%)", y = "Density",  title = "Lake Mendota") + theme(legend.title = element_blank())

# spark_within <- within_reps(snorm)
# spark_within <- apply(spark_within, 1, mean, na.rm = T)
# spark_across <- apply(snorm, 1, cv, na.rm = T)
# cv_stat <- c(rep("within replicates", length(spark_within)), rep("across samples", length(spark_across)))
# spark_cv_stats <- data.frame(type = cv_stat, stat = c(spark_within, spark_across))
# 
# p2 <- ggplot(spark_cv_stats, aes(stat, fill = type)) + geom_density(alpha = 0.5) + scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + labs(x = "Coefficent of Variance (%)", title = "Sparkling Lake") + theme(legend.title = element_blank())
# 
# 
# trout_within <- within_reps(tnorm)
# trout_within <- apply(trout_within, 1, mean, na.rm = T)
# trout_across <- apply(tnorm, 1, cv, na.rm = T)
# cv_stat <- c(rep("within replicates", length(trout_within)), rep("across samples", length(trout_across)))
# trout_cv_stats <- data.frame(type = cv_stat, stat = c(trout_within, trout_across))
# 
# p3 <- ggplot(trout_cv_stats, aes(stat, fill = type)) + geom_density(alpha = 0.5) + scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + labs(x = "Coefficent of Variance (%)", title = "Trout Bog") + theme(legend.title = element_blank())
# 
# plot_grid(p1, p2, p3, nrow = 3, labels = c("A", "B", "C"))

# Plot example trends
mnorm2 <- mnorm[1:1000,]
mnorm2$Genes <- rownames(mnorm2)
mnorm2 <- melt(mnorm2)
mnorm2$variable <- gsub(".nonrRNA", "", mnorm2$variable)
mnorm2$Time <- metadata$Timepoint[match(mnorm2$variable, metadata$Sample)]
agg_mnorm <- aggregate(value ~ Time + Genes, mnorm2, mean)
agg_mnorm$Time <- factor(agg_mnorm$Time, levels = c("0", "4", "8", "12", "16", "20", "24", "28", "32", "36", "40", "44"))

genes <- unique(agg_mnorm$Genes)
example <- data.frame("time" = agg_mnorm$Time[which(agg_mnorm$Genes == genes[1])], "eigenvector" = zscore(agg_mnorm$value[which(agg_mnorm$Genes == genes[32])]))

# Gene 6 is up down
# Gene 7 is on once only
# Gene 32 is day one


p2 <- ggplot(data = example, aes(x = time, y = eigenvector)) + geom_bar(stat = "identity") + labs(x = " ", y = " ")
example <- data.frame("time" = agg_mnorm$Time[which(agg_mnorm$Genes == genes[1])], "eigenvector" = zscore(agg_mnorm$value[which(agg_mnorm$Genes == genes[6])]))
p3 <- ggplot(data = example, aes(x = time, y = eigenvector)) + geom_bar(stat = "identity") + labs(x = " ", y = "Eigenvector of gene expression")
example <- data.frame("time" = agg_mnorm$Time[which(agg_mnorm$Genes == genes[1])], "eigenvector" = zscore(agg_mnorm$value[which(agg_mnorm$Genes == genes[7])]))
p4 <- ggplot(data = example, aes(x = time, y = eigenvector)) + geom_bar(stat = "identity") + labs(x = "Hours since start of time series", y = " ")

part2 <- plot_grid(p2, p3, p4, nrow = 3)
plot_grid(p1, part2, nrow = 1, labels = c("A", "B"))

```




**Supp figure. Assessing the variability of metatranscriptomic read counts.** One aim of this metatranscriptomic study was to provide information on the variability in gene expression in freshwater that could be used to guide further metatranscriptomic experiments. We calculated the coefficient of variance (CoV) for each gene both within replicates and across samples (A). CoV was lower within replicates than across replicates, indicating that variation across replicates is not technical. High levels of variability were observed in genes from all three lakes. In panel B, three example eigenvectors of genes are shown, including trends for 1) on one day, off the next,  2) up and down across the time series, and 3) observed in only one timepoint. 


