---
title: "GEODES"
author: "Alex Linz"
date: "September 28, 2017"
output: html_document
---

#### Diel transcriptomics of freshwater lakes

GEODES stands for "Gene Expression in Oligotrophic, Dystrophic, and Eutrophic Systems."
In this experiment, we collected RNA from three lakes every 4 hours for 48 hours, with the goal of analyzing diel trends.
The aims of this project are:

- Identify interspecies(taxa) carbon compound exchanges, particularly between phototrophic and heterotrophic bacteria
- Identify carbon compounds used by the community that have no standing pool in the environment (cryptic compounds)
- Compare RNA to DNA to find out what proportion of the community is active, and what proportion of their genomes are active
- Generate new reference genomes for freshwater
- Identify new pathways and potential functions based on co-expression of genes
- Look at genes of interest from previous studies, such as actinorhodopsin

Copyright Katherine McMahon, Alex Linz, and friends (2017)

Want more background about GEODES? Check out these blog posts! https://uwmcmahonlab.wordpress.com/2016/05/04/introducing-geodes/ & https://uwmcmahonlab.wordpress.com/2016/11/14/geodes-data-preview/


####Where do I get the data?

Are you in the McMahon Lab or one of our collaborators?

- Go to /lakes_data/Metatranscriptomes/geodes/ on our storage drive.
- If you can't access the storage drive, see this page: https://github.com/McMahonLab/McMahon_Fileshare

Not in the lab but curious about our data?

- Download our data from http://genome.jgi.doe.gov/Diecycetabolisms/Diecycetabolisms.info.html . Most metatranscriptomes became public on Aug 1, 2017, but some samples are earlier and others later.
- Shoot us an email! trina.mcmahon@wisc.edu or amlinz@wisc.edu

__BE WARNED__ We're not making the raw data or intermediate files available on Github because it would far exceed our space limit. The quality filtered, compressed metatranscriptome files total 185GB! Download at your own risk.

####Repo Structure

- protocols
	- Field and lab protocols go here.
- environmental_data
	- for_database
		- Files formatted for upload to the McMahon Lab SQL database
	- for_humans
		- Excel files that are easily interpreted by people
- bioinformatics_workflow
  - executable_files
    - Bash scripts that actually process the data
  - lab_notebooks
    - My notes from writing this scripts. May contain sass.
  - programs
    - Pre-built tarballs of the programs used for use on CHTC
  - R_processing
    - Table formatting once the data was small enough to load into R
  - submit_files
    - Tells CHTC how to run your jobs
  - submit_node_scripts
    - Extremely minor tasks to run on the submit node itself that need to run between parallel tasks - like making a list of files to queue or moving files around
    - Full_GEODES_Workflow.Rmd/.html
      - 1st attempt at workflow. Now out of date but has some useful info on why I chose to do things the way I did. Will be updated on 2nd interation of workflow
    - GEODES_bioinformatics_instructions.Rmd
      - How to run the current workflow on CHTC