# Process mapped reads

####Goal of this analysis

The mapping in the last step produced a .sam file where each line is a read in the metatranscriptome, and information is included about whether or not it mapped, and where it mapped. Technically, this is the data I need to address my hypotheses, but each file is way to large and clunky to analyze. I'd like to analyze these using the DESEQ R package, which requires a table where rows are genes and columns are samples. 

##### Approach

I'm going to use htseq to count up to reads mapping to the same thing in each metaT. I'll need samtools again to reformat the mapping output. Once I get an htseq-count summary file for each sample, I'll combine them into a table for DESEQ. While I'm here, I'll also create metadata files from the gff file (reproducibly!) so I can redo it for any genomes in my database.

## Most recent workflow. 
####Use this if you want to replicate our protocol. Updated ...

Set up folders in gluster for the outputs:
```{r, eval = F}
mkdir /mnt/gluster/amlinz/GEODES_mapping_summaries/
```

Build python install
interactive.sub:
```{r, eval = F}
universe = vanilla
# Name the log file:
log = interactive.log

# Name the files where standard output and error should be saved:
output = process.out
error = process.err

# If you wish to compile code, you'll need the below lines. 
#  Otherwise, LEAVE THEM OUT if you just want to interactively test!
+IsBuildJob = true
requirements = (OpSysAndVer =?= "SL6") && ( IsBuildSlot == true )

# Indicate all files that need to go into the interactive job session,
#  including any tar files that you prepared:
transfer_input_files = Python-2.7.13.tgz

# It's still important to request enough computing resources. The below 
#  values are a good starting point, but consider your file sizes for an
#  estimate of "disk" and use any other information you might have
#  for "memory" and/or "cpus".
request_cpus = 1
request_memory = 1GB
request_disk = 1GB

queue
```

Start the session with:
```{r, eval = F}
condor_submit -i interactive.sub
```

Here's what I'm typing in the interactive session:
```{r, eval = F}
mkdir python
tar -xvf Python-2.7.13.tgz
cd Python-2.7.13
./configure --prefix=$(pwd)/../python
make
make install
cd ..
ls python
ls python/bin
# Maybe I can install htseq right in here?
export PATH=$(pwd)/python/bin:$PATH
wget https://bootstrap.pypa.io/get-pip.py
python get-pip.py
pip install numpy
pip install matplotlib
pip install htseq
pip install pysam
# hey, that worked!
tar -czvf python.tar.gz python/
exit

# In your home folder, copy the install to gluster (it's too big to transfer via the submit file)
cp python.tar.gz /mnt/gluster/amlinz/
```

Make a list of files to run:
```{r, eval = F}
for file in /mnt/gluster/amlinz/GEODES_mapping_concat/*; do sample=$(basename $file |cut -d'.' -f1); echo $sample;done > samfiles.txt
cat samfiles.txt

```


#Lab Notebook

####2017-02-21

Some of this is carried over from the 02mapping lab notebook, since I was originally planning to do that in one step. However, the script was large, slow and error prone, so here we are. I'll need my samtools install from the last script, and the python install with the htseq module.

Make a list of files to run:
```{r, eval = F}
ls /mnt/gluster/amlinz/GEODES_mapping_concat/ > samfiles.txt
cat samfiles.txt

```

Copy the python file over to gluster then write submit script - the only things it's sending over is the samtools install and the gff file.
03processing.sub:
```{r, eval = F}
# 03processing.sub
#
#
# Specify the HTCondor Universe
universe = vanilla
log = 03processing_$(Cluster).log
error = 03processing_$(Cluster)_$(Process).err
requirements = (OpSys == "LINUX") && (OpSysMajorVer == 6)
#
# Specify your executable, arguments, and a file for HTCondor to store standard
#  output.
executable = 03processing.sh
arguments = $(samplename)
output = 03processing_$(Cluster).out
#
# Specify that HTCondor should transfer files to and from the
#  computer where each job runs.
should_transfer_files = YES
when_to_transfer_output = ON_EXIT
transfer_input_files = samtools.tar.gz,mapping_database.gff.gz
#transfer_output_files =
#
# Tell HTCondor what amount of compute resources
#  each job will need on the computer where it runs.
Requirements = (Target.HasGluster == true)
request_cpus = 1
request_memory = 3GB
request_disk = 5GB
#
# Tell HTCondor to run every fastq file in the provided list:
queue samplename from samfiles.txt
```

And the executable, based on my early mapping testing with htseq:
```{r, eval = F}
#!/bin/bash
#Make a count table of mapped reads

#Transfer data from gluster
cp /mnt/gluster/amlinz/GEODES_mapping_concat/$1.all.sam .
cp /mnt/gluster/amlinz/python.tar.gz .


#Unzip programs
tar xzf python.tar.gz
tar xvf samtools.tar.gz
gzip -d mapping_database.gff.gz

#Update the path variable
mkdir home
export PATH=$(pwd)/python/bin:$(pwd)/samtools/bin:$PATH
export HOME=$(pwd)/home

#Manipulate the output
samtools view -o $1.bam $1.all.sam
samtools sort $1.bam $1.sorted
samtools index $1.sorted.bam

#Count reads
python ./python/bin/htseq-count -f bam -r pos -s no -a 0 -t CDS -i locus_tag -m intersection-strict -o $1.CDS.sam $1.sorted.bam mapping_database.gff > $1.CDS.out

#Save output to gluster. Using cp/rm instead of mv so it will overwrite old output in gluster.
cp $1.CDS.out /mnt/gluster/amlinz/GEODES_mapping_summaries/
rm $1.CDS.out

#Clean up
rm -rf python
rm -rf samtools
rm *.sam
rm *.bam
rm *.bai
rm mapping_database*
rm *.tar.gz
  
```

Something's wrong with the header in the all.sam file? That would be an issue with the merging script. Going into interactive mode to check. samtools is a pain and their documentation seems out of date.

Fixed the issue - left out the -h flag in samtools view that leaves the header in output. Rerun 02merge script and 03processing, and got the following error:

Error occured when processing SAM input (record #285895 in file GEODES002_nonrRNA.all.sam):
  object of type 'NoneType' has no len()
  [Exception type: TypeError, raised in _HTSeq.pyx:771]

It happened in all three files, although in different lines. The only thing I can find online is here? https://sourceforge.net/p/htseq/bugs/12/ Doesn't seem helpful because many of my lines have *. Will sleep on this and come back tomorrow.

