# Add phylum-level gene classifications to keys when genome classifications are not available

# Read in gene keys
path = "/Users/Alex/Desktop/geodes_data_tables/"

mendota_key <- read.csv(paste(path, "Mendota_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)
trout_key <- read.csv(paste(path, "Trout_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)
spark_key <- read.csv(paste(path, "Sparkling_ID90_genekey_reclassified_2018-03-03.csv", sep = ""), header = T)


# Read in phylodist files

disk = "/Volumes/Alex Linz Thesis Data/GEODES/GEODES_metaG_assemblies/phylogeny/"
GEODES005 <- read.delim(paste(disk, "GEODES005.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES006 <- read.delim(paste(disk, "GEODES006.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES057 <- read.delim(paste(disk, "GEODES057.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES058 <- read.delim(paste(disk, "GEODES058.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES117 <- read.delim(paste(disk, "GEODES117.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES118 <- read.delim(paste(disk, "GEODES118.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES165 <- read.delim(paste(disk, "GEODES165.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES166 <- read.delim(paste(disk, "GEODES166.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES167 <- read.delim(paste(disk, "GEODES167.assembled.phylodist", sep = ""), sep = "\t", header = F)
GEODES168 <- read.delim(paste(disk, "GEODES168.assembled.phylodist", sep = ""), sep = "\t", header = F)

# Concatentae phylodist files from the same lakes and remove the originals to save space

mendota_phylodist <- rbind(GEODES117, GEODES118)
rm(GEODES117)
rm(GEODES118)

trout_phylodist <- rbind(GEODES057, GEODES058)
rm(GEODES057)
rm(GEODES058)

spark_phylodist <- rbind(GEODES005, GEODES006, GEODES165, GEODES166, GEODES167, GEODES168)
rm(GEODES005)
rm(GEODES006)
rm(GEODES165)
rm(GEODES166)
rm(GEODES167)
rm(GEODES168)

# For each completely unclassified gene in the key files, see if it is in the phylodist files and grab the phylum classification of the taxonomy string in the phylodist file

mendota_key$Taxonomy <- gsub("Bacteria;", "", mendota_key$Taxonomy)
mendota_key$Taxonomy <- gsub("Proteobacteria;", "", mendota_key$Taxonomy)
mendota_key$Taxonomy <- gsub("Eukaryota;", "", mendota_key$Taxonomy)
mendota_key$Phylum <- sapply(strsplit(as.character(mendota_key$Taxonomy),";"), `[`, 1)

mendota_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("None", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("unclassified", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Unclassified ", "Unclassified", mendota_key$Phylum)
mendota_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", mendota_key$Phylum)
mendota_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", mendota_key$Phylum)
mendota_key$Phylum[grep("Blank", mendota_key$Phylum)] <- "Unclassified"
mendota_key$Phylum[which(is.na(mendota_key$Phylum) == T)] <- "Unclassified"

unclassified <- which(mendota_key$Phylum == "Unclassified")
unclassified_genes <- as.character(mendota_key$Gene[unclassified])
unclassified_genes <- gsub("_GEODES117", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES118", "", unclassified_genes)
tax_string <- mendota_phylodist$V5[match(unclassified_genes, mendota_phylodist$V1)]
tax_string <- gsub("Bacteria;", "", tax_string)
tax_string <- gsub("Archaea;", "", tax_string)
tax_string <- gsub("Bacteria;Proteobacteria;", "", tax_string)
tax_string <- gsub("Proteobacteria;", "", tax_string)
tax_string <- sapply(strsplit(as.character(tax_string),";"), `[`, 1)
tax_string <- gsub("unclassified ", "", tax_string)                     
tax_string <- gsub("Coprobacter", "Bacteroidetes", tax_string)   
tax_string <- gsub("Bankia setacea gill symbiont BS12", "Gammaproteobacteria", tax_string)   
tax_string <- gsub("Betaproteobacteria bacterium AAP51", "Betaproteobacteria", tax_string) 
tax_string <- gsub("Candidatus Pelagibacter", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Afipia", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Synechococcus", "Cyanobacteria", tax_string)
tax_string <- gsub("Cyclobacteriaceae bacterium AK24", "Bacteroidetes", tax_string)
tax_string <- gsub("Candidatus ", "", tax_string)
tax_string <- gsub("Rhizobium", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Methyloversatilis", "Betaproteobacteria", tax_string)
tax_string <- gsub("Verrucomicrobium", "Verrucomicrobia", tax_string)
tax_string <- gsub("Burkholderia", "Betaproteobacteria", tax_string)
tax_string <- gsub("Lysinimicrobium", "Actinobacteria", tax_string)
tax_string <- gsub("Microvirga", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Vampirovibrio", "Cyanobacteria", tax_string)
tax_string <- gsub("Haloplasmatales", "Tenericutes", tax_string)
tax_string <- gsub("Listeriaceae bacterium FSL A5-0281", "Firmicutes", tax_string)
tax_string <- gsub("Cedecea", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Haemophilus", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Porphyromonadaceae bacterium ING2-E5B", "Bacteroidetes", tax_string)
tax_string <- gsub("Gordonia", "Actinobacteria", tax_string)
tax_string <- gsub("Sneathia", "Fusobacteria", tax_string)
tax_string <- gsub("Campylobacter", "Epsilonproteobacteria", tax_string)
tax_string <- gsub("Lachnospiraceae bacterium 40969", "Firmicutes", tax_string)
tax_string <- gsub("Saccharomonospora", "Actinobacteria", tax_string)
tax_string <- gsub("Bilophila", "Deltaproteobacteria", tax_string)
tax_string <- gsub("Bacillus", "Firmicutes", tax_string)
tax_string <- gsub("Lactobacillus", "Firmicutes", tax_string)
tax_string <- gsub("Campylobacterales sp. GD 1", "Epsilonproteobacteria", tax_string)
tax_string <- gsub("Clostridium", "Firmicutes", tax_string)
tax_string <- gsub("Propionibacterium", "Actinobacteria", tax_string)
tax_string <- gsub("Pseudomonas", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Rhodococcus", "Actinobacteria", tax_string)
tax_string <- gsub("Vibrio", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("EM3", "Hydrothermae", tax_string) 
tax_string <- gsub("Bacteroides", "Bacteroidetes", tax_string)   
tax_string <- gsub("Salmonella", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Epsilonproteobacteriaales sp. GD 1", "Epsilonproteobacteria", tax_string) 
tax_string <- gsub("Halobacteriaceae archaeon HSR2", "Euryarchaeota", tax_string) 
tax_string <- gsub("Methanobacterium", "Euryarchaeota", tax_string) 

tax_string[which(is.na(tax_string) == T)] <- "Unclassified"

mendota_key$Phylum[unclassified] <- tax_string

write.csv(mendota_key, paste(path, "Mendota_ID90_genekey_geneclassifications_2018-11-28.csv", sep = ""), row.names = F)

# Repeat with Trout

trout_key$Taxonomy <- gsub("Bacteria;", "", trout_key$Taxonomy)
trout_key$Taxonomy <- gsub("Proteobacteria;", "", trout_key$Taxonomy)
trout_key$Taxonomy <- gsub("Eukaryota;", "", trout_key$Taxonomy)
trout_key$Phylum <- sapply(strsplit(as.character(trout_key$Taxonomy),";"), `[`, 1)

trout_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("None", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified unclassified", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("unclassified", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("Unclassified ", "Unclassified", trout_key$Phylum)
trout_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", trout_key$Phylum)
trout_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", trout_key$Phylum)
trout_key$Phylum[grep("Blank", trout_key$Phylum)] <- "Unclassified"

unclassified <- which(trout_key$Phylum == "Unclassified")
unclassified_genes <- as.character(trout_key$Gene[unclassified])
unclassified_genes <- gsub("_GEODES057", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES058", "", unclassified_genes)
tax_string <- trout_phylodist$V5[match(unclassified_genes, trout_phylodist$V1)]
tax_string <- gsub("Bacteria;", "", tax_string)
tax_string <- gsub("Archaea;", "", tax_string)
tax_string <- gsub("Bacteria;Proteobacteria;", "", tax_string)
tax_string <- gsub("Proteobacteria;", "", tax_string)
tax_string <- sapply(strsplit(as.character(tax_string),";"), `[`, 1)
tax_string <- gsub("unclassified ", "", tax_string)                     
tax_string <- gsub("Coprobacter", "Bacteroidetes", tax_string)   
tax_string <- gsub("Bankia setacea gill symbiont BS12", "Gammaproteobacteria", tax_string)   
tax_string <- gsub("Betaproteobacteria bacterium AAP51", "Betaproteobacteria", tax_string) 
tax_string <- gsub("Candidatus Pelagibacter", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Afipia", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Synechococcus", "Cyanobacteria", tax_string)
tax_string <- gsub("Cyclobacteriaceae bacterium AK24", "Bacteroidetes", tax_string)
tax_string <- gsub("Candidatus ", "", tax_string)
tax_string <- gsub("Rhizobium", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Methyloversatilis", "Betaproteobacteria", tax_string)
tax_string <- gsub("Verrucomicrobium", "Verrucomicrobia", tax_string)
tax_string <- gsub("Burkholderia", "Betaproteobacteria", tax_string)
tax_string <- gsub("Lysinimicrobium", "Actinobacteria", tax_string)
tax_string <- gsub("Microvirga", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Vampirovibrio", "Cyanobacteria", tax_string)
tax_string <- gsub("Haloplasmatales", "Tenericutes", tax_string)
tax_string <- gsub("Listeriaceae bacterium FSL A5-0281", "Firmicutes", tax_string)
tax_string <- gsub("Cedecea", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Haemophilus", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Porphyromonadaceae bacterium ING2-E5B", "Bacteroidetes", tax_string)
tax_string <- gsub("Gordonia", "Actinobacteria", tax_string)
tax_string <- gsub("Sneathia", "Fusobacteria", tax_string)
tax_string <- gsub("Campylobacter", "Epsilonproteobacteria", tax_string)
tax_string <- gsub("Lachnospiraceae bacterium 40969", "Firmicutes", tax_string)
tax_string <- gsub("Saccharomonospora", "Actinobacteria", tax_string)
tax_string <- gsub("Bilophila", "Deltaproteobacteria", tax_string)
tax_string <- gsub("Bacillus", "Firmicutes", tax_string)
tax_string <- gsub("Lactobacillus", "Firmicutes", tax_string)
tax_string <- gsub("Campylobacterales sp. GD 1", "Epsilonproteobacteria", tax_string)
tax_string <- gsub("Clostridium", "Firmicutes", tax_string)
tax_string <- gsub("Propionibacterium", "Actinobacteria", tax_string)
tax_string <- gsub("Pseudomonas", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Rhodococcus", "Actinobacteria", tax_string)
tax_string <- gsub("Vibrio", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("EM3", "Hydrothermae", tax_string) 
tax_string <- gsub("Bacteroides", "Bacteroidetes", tax_string)   
tax_string <- gsub("Salmonella", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Epsilonproteobacteriaales sp. GD 1", "Epsilonproteobacteria", tax_string) 
tax_string <- gsub("Halobacteriaceae archaeon HSR2", "Euryarchaeota", tax_string) 
tax_string <- gsub("Methanobacterium", "Euryarchaeota", tax_string) 

tax_string[which(is.na(tax_string) == T)] <- "Unclassified"

trout_key$Phylum[unclassified] <- tax_string
write.csv(trout_key, paste(path, "Trout_ID90_genekey_geneclassifications_2018-11-28.csv", sep = ""), row.names = F)

# Lastly, Sparkling
spark_key$Taxonomy <- gsub("Bacteria;", "", spark_key$Taxonomy)
spark_key$Taxonomy <- gsub("Proteobacteria;", "", spark_key$Taxonomy)
spark_key$Taxonomy <- gsub("Eukaryota;", "", spark_key$Taxonomy)
spark_key$Phylum <- sapply(strsplit(as.character(spark_key$Taxonomy),";"), `[`, 1)

spark_key$Phylum <- gsub("Cryptophyta,Cryptophyceae,Pyrenomonadales,Geminigeraceae,Guillardia,theta", "Cryptophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("Haptophyta,Prymnesiophyceae,Isochrysidales,Noelaerhabdaceae,Emiliania,huxleyi", "Haptophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Coscinodiscophyceae,Thalassiosirales,Thalassiosiraceae,Thalassiosira,pseudonana", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Pelagophyceae,Pelagomonadales,Pelagomonadaceae,Aureococcus,anophagefferens", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Bacillariophyceae,Naviculales,Phaeodactylaceae,Phaeodactylum,tricornutum", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("Heterokonta,Ochrophyta,Eustigmataphyceae,Eustigmataceae,Nannochloropsis,gaditana", "Heterokonta", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified unclassified unclassified unclassified unclassified", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION MH", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION LP", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION DUE TO FEW HITS IN PHYLODIST", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("NO CLASSIFICATION BASED ON GIVEN PHYLODIST", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("None", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified unclassified Perkinsida", "Perkinsozoa", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified unclassified", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified Oligohymenophorea", "Ciliophora", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified Pelagophyceae", "Ochrophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("unclassified", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("Unclassified ", "Unclassified", spark_key$Phylum)
spark_key$Phylum <- gsub("UnclassifiedIsochrysidales", "Haptophyta", spark_key$Phylum)
spark_key$Phylum <- gsub("Crenarchaeaota", "Crenarchaeota", spark_key$Phylum)
spark_key$Phylum[grep("Blank", spark_key$Phylum)] <- "Unclassified"

unclassified <- which(spark_key$Phylum == "Unclassified")
unclassified_genes <- as.character(spark_key$Gene[unclassified])
unclassified_genes <- gsub("_GEODES005", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES006", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES165", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES166", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES167", "", unclassified_genes)
unclassified_genes <- gsub("_GEODES168", "", unclassified_genes)
tax_string <- spark_phylodist$V5[match(unclassified_genes, spark_phylodist$V1)]
tax_string <- gsub("Bacteria;", "", tax_string)
tax_string <- gsub("Archaea;", "", tax_string)
tax_string <- gsub("Bacteria;Proteobacteria;", "", tax_string)
tax_string <- gsub("Proteobacteria;", "", tax_string)
tax_string <- sapply(strsplit(as.character(tax_string),";"), `[`, 1)
tax_string <- gsub("unclassified ", "", tax_string)                     
tax_string <- gsub("Coprobacter", "Bacteroidetes", tax_string)   
tax_string <- gsub("Bankia setacea gill symbiont BS12", "Gammaproteobacteria", tax_string)   
tax_string <- gsub("Betaproteobacteria bacterium AAP51", "Betaproteobacteria", tax_string) 
tax_string <- gsub("Candidatus Pelagibacter", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Afipia", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Synechococcus", "Cyanobacteria", tax_string)
tax_string <- gsub("Cyclobacteriaceae bacterium AK24", "Bacteroidetes", tax_string)
tax_string <- gsub("Candidatus ", "", tax_string)
tax_string <- gsub("Rhizobium", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Methyloversatilis", "Betaproteobacteria", tax_string)
tax_string <- gsub("Verrucomicrobium", "Verrucomicrobia", tax_string)
tax_string <- gsub("Burkholderia", "Betaproteobacteria", tax_string)
tax_string <- gsub("Lysinimicrobium", "Actinobacteria", tax_string)
tax_string <- gsub("Microvirga", "Alphaproteobacteria", tax_string)
tax_string <- gsub("Vampirovibrio", "Cyanobacteria", tax_string)
tax_string <- gsub("Haloplasmatales", "Tenericutes", tax_string)
tax_string <- gsub("Listeriaceae bacterium FSL A5-0281", "Firmicutes", tax_string)
tax_string <- gsub("Cedecea", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Haemophilus", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Porphyromonadaceae bacterium ING2-E5B", "Bacteroidetes", tax_string)
tax_string <- gsub("Gordonia", "Actinobacteria", tax_string)
tax_string <- gsub("Sneathia", "Fusobacteria", tax_string)
tax_string <- gsub("Campylobacter", "Epsilonproteobacteria", tax_string)
tax_string <- gsub("Lachnospiraceae bacterium 40969", "Firmicutes", tax_string)
tax_string <- gsub("Saccharomonospora", "Actinobacteria", tax_string)
tax_string <- gsub("Bilophila", "Deltaproteobacteria", tax_string)
tax_string <- gsub("Bacillus", "Firmicutes", tax_string)
tax_string <- gsub("Lactobacillus", "Firmicutes", tax_string)
tax_string <- gsub("Campylobacterales sp. GD 1", "Epsilonproteobacteria", tax_string)
tax_string <- gsub("Clostridium", "Firmicutes", tax_string)
tax_string <- gsub("Propionibacterium", "Actinobacteria", tax_string)
tax_string <- gsub("Pseudomonas", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Rhodococcus", "Actinobacteria", tax_string)
tax_string <- gsub("Vibrio", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("EM3", "Hydrothermae", tax_string) 
tax_string <- gsub("Bacteroides", "Bacteroidetes", tax_string)   
tax_string <- gsub("Salmonella", "Gammaproteobacteria", tax_string) 
tax_string <- gsub("Epsilonproteobacteriaales sp. GD 1", "Epsilonproteobacteria", tax_string) 
tax_string <- gsub("Halobacteriaceae archaeon HSR2", "Euryarchaeota", tax_string) 
tax_string <- gsub("Methanobacterium", "Euryarchaeota", tax_string) 

tax_string[which(is.na(tax_string) == T)] <- "Unclassified"

spark_key$Phylum[unclassified] <- tax_string
write.csv(spark_key, paste(path, "Sparkling_ID90_genekey_geneclassifications_2018-11-28.csv", sep = ""), row.names = F)