---
title: "Genome-centric analyses"
author: "Alex Linz"
date: "August 23, 2018"
output: html_document
---

```{r, setup, eval = T, echo = F, warning = F, message = F}
# Packages:
library(reshape2)
library(ggplot2)
library(cowplot)

# Read in data
readcounts <- read.csv("C:/Users/Goose and Gander/Documents/geodes_data_tables/GEODES_genomes_normalized.csv", header = T, row.names = 1)
genekey <- read.csv("C:/Users/Goose and Gander/Documents/geodes_data_tables/GEODES_genomes_genekey.csv", header = T, colClasses = c("character"))
metadata <- read.csv(file = "C:/Users/Goose and Gander/Desktop/geodes/bioinformatics_workflow/R_processing/sample_metadata.csv", header = T)
genomekey <- read.csv("C:/Users/Goose and Gander/Dropbox/genome_quality.csv", header = T, colClasses = c("character"))

# Some light data processing
# There's an extra column of outdated row numbers in genekey - remove
genekey$X <- NULL
# Swap contig names for bin names
swap <- grep("binned", genekey$Geneid)
split_names <- sapply(strsplit(genekey$Geneid,"_"), `[`, 3)
genekey$Genome[swap] <- split_names[swap]


# I know from my gene-centric analyses that there are several samples with abnormal results. Remove these:
colnames(readcounts) <- gsub(".nonrRNA", "", colnames(readcounts))
readcounts <- readcounts[, which(colnames(readcounts) != "GEODES014" & colnames(readcounts) != "GEODES033" & colnames(readcounts) != "GEODES065" & colnames(readcounts) != "GEODES158" & colnames(readcounts) != "GEODES053")]

```


```{r, rank_abundance, eval = T, echo = F, message = F, warning = F, fig.width = 12, fig.height = 4}
readcounts$Genome <- sapply(strsplit(rownames(readcounts),"_"), `[`, 3)
readcounts2 <- melt(readcounts)
readcounts2$Lake <- metadata$Lake[match(readcounts2$variable, metadata$Sample)]

genomecounts <- aggregate(value ~ Genome + Lake, readcounts2, sum)
genomecounts$Phylum <- genomekey$Phylum[match(genomecounts$Genome, genomekey$Genome.Name)]

MEcounts <- genomecounts[which(genomecounts$Lake == "Mendota"), ]
SPcounts <- genomecounts[which(genomecounts$Lake == "Sparkling"), ]
TBcounts <- genomecounts[which(genomecounts$Lake == "Trout"), ]

MEcounts <- MEcounts[which(MEcounts$Phylum != "transcript"), ]
SPcounts <- SPcounts[which(SPcounts$Phylum != "transcript"), ]
TBcounts <- TBcounts[which(TBcounts$Phylum != "transcript"), ]

# Normalize reads by genome size
MEcounts$Lengths <- as.numeric(genomekey$Length[match(MEcounts$Genome, genomekey$Genome.Name)])
SPcounts$Lengths <- as.numeric(genomekey$Length[match(SPcounts$Genome, genomekey$Genome.Name)])
TBcounts$Lengths <- as.numeric(genomekey$Length[match(TBcounts$Genome, genomekey$Genome.Name)])

MEcounts$adj.value <- MEcounts$value/(MEcounts$Length/1000000)
SPcounts$adj.value <- SPcounts$value/(SPcounts$Length/1000000)
TBcounts$adj.value <- TBcounts$value/(TBcounts$Length/1000000)

# order by expression amount
MEcounts <- MEcounts[order(MEcounts$adj.value, decreasing = T), ]
SPcounts <- SPcounts[order(SPcounts$adj.value, decreasing = T), ]
TBcounts <- TBcounts[order(TBcounts$adj.value, decreasing = T), ]

MEcounts$Genome <- factor(MEcounts$Genome, levels = MEcounts$Genome)
SPcounts$Genome <- factor(SPcounts$Genome, levels = SPcounts$Genome)
TBcounts$Genome <- factor(TBcounts$Genome, levels = TBcounts$Genome)

ggplot(MEcounts[1:50,], aes(x = Genome, y = adj.value, fill = Phylum)) + geom_col() + labs(x = NULL, y = "Sum Transcripts/L per 1 million bp", title = "Mendota") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_fill_brewer(palette = "Paired")

ggplot(SPcounts[1:50,], aes(x = Genome, y = adj.value, fill = Phylum)) + geom_col() + labs(x = NULL, y = "Sum Transcripts/L per 1 million bp", title = "Sparkling") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_fill_brewer(palette = "Paired")

ggplot(TBcounts[1:50,], aes(x = Genome, y = adj.value, fill = Phylum)) + geom_col() + labs(x = NULL, y = "Sum Transcripts/L per 1 million bp", title = "Trout") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_fill_brewer(palette = "Paired")

```


```{r, top10, eval = T, echo = F}
print("Mendota:")
print(genomekey[match(MEcounts$Genome[1:10], genomekey$Genome.Name), ])
print("Sparkling:")
print(genomekey[match(SPcounts$Genome[1:10], genomekey$Genome.Name), ])
print("Trout:")
print(genomekey[match(TBcounts$Genome[1:10], genomekey$Genome.Name), ])
```


```{r, novosphingobium, eval = F, echo = F}

lakekey <- metadata$Lake[match(colnames(readcounts), metadata$Sample)]
genome <- "GEODES057-binned.063"
genes <- genekey[which(genekey$Genome == genome), ]
reads <- readcounts[match(genes$Geneid, rownames(readcounts)), which(lakekey == "Trout")]
reads <- reads[which(rowSums(reads) > 0), ]

cor.data <- expand.grid(rownames(reads), rownames(reads))
cor.data <- cor.data[which((cor.data$Var1 == cor.data$Var2) == F),]
test.dups <- duplicated(t(apply(cor.data, 1, sort)))
cor.data <- cor.data[which(test.dups == F), ]

correlations <- c()
for(i in 1:dim(cor.data)[1]){
  thing1 <- reads[which(rownames(reads) == cor.data$Var1[i]), ]
  thing2 <- reads[which(rownames(reads) == cor.data$Var2[i]), ]
  correlations[i] <- cor(as.numeric(thing1), as.numeric(thing2))
  if(abs(correlations[i]) < 0.95){
    correlations[i] <- 0
  }
}
cor.data$edges <- correlations
cor.data <- cor.data[which(cor.data$edges != 0), ]

net <- graph_from_data_frame(cor.data, directed = F)
l <- layout_with_lgl(net)
plot(net, vertex.color = "white", )
ceb <- cluster_edge_betweenness(net) 
plot(ceb, net, vertex.color="lightgrey", vertex.size = V(net)$size, vertex.label = membership(ceb))

clusters <- membership(ceb)

genekey$Product[match(names(clusters)[which(clusters == 85)], genekey$Geneid)]

```